// @fundamental-ngx/core/esm2015/lib/notification/notification-header/notification-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, HostBinding, Input, Output, ViewEncapsulation } from '@angular/core';
export class NotificationHeaderComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdNotificationHeaderClass = true;
        /**
         * Whether user wants to hide close button at the top. It's shown by default.
         */
        this.closeButton = true;
        /**
         * Event thrown always, when the close button is clicked
         */
        this.closeButtonClick = new EventEmitter();
    }
    /**
     * @hidden
     * @return {?}
     */
    closeButtonClicked() {
        this.closeButtonClick.emit();
    }
}
NotificationHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-notification-header',
                template: "<div [ngClass]=\"['fd-notification__indicator--' + type]\" *ngIf=\"type\"></div>\n<ng-content select=\"[fd-notification-title]\"></ng-content>\n<button fd-button [glyph]=\"'decline'\" [options]=\"'light'\" class=\"fd-notification__close\" (click)=\"closeButtonClicked()\"></button>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [""]
            }] }
];
NotificationHeaderComponent.propDecorators = {
    fdNotificationHeaderClass: [{ type: HostBinding, args: ['class.fd-notification__header',] }],
    type: [{ type: Input }],
    closeButton: [{ type: Input }],
    closeButtonClick: [{ type: Output }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationHeaderComponent.prototype.fdNotificationHeaderClass;
    /**
     * type of Notification 'success' | 'warning' | 'information' | 'error'
     * @type {?}
     */
    NotificationHeaderComponent.prototype.type;
    /**
     * Whether user wants to hide close button at the top. It's shown by default.
     * @type {?}
     */
    NotificationHeaderComponent.prototype.closeButton;
    /**
     * Event thrown always, when the close button is clicked
     * @type {?}
     */
    NotificationHeaderComponent.prototype.closeButtonClick;
}
