// @smart/utils/esm2015/services/rest/rest-service.factory.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
var RestServiceFactory_1;
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { RestClient } from './rest-client';
import { LogService } from '../log.service';
/** @internal */
let RestServiceFactory = RestServiceFactory_1 = class RestServiceFactory {
    constructor(httpClient) {
        this.httpClient = httpClient;
        this.map = new Map();
        this.basePath = null;
        this.DOMAIN = null;
        this.IDENTIFIER = 'identifier';
    }
    static setGlobalBasePath(globalBasePath) {
        if (!RestServiceFactory_1.globalBasePath) {
            RestServiceFactory_1.globalBasePath = globalBasePath;
        }
        else {
            RestServiceFactory_1.logService.warn('The value of a global base path was already set. ' +
                'Update is not possible, the value remained unchanged!');
        }
    }
    static getGlobalBasePath() {
        return RestServiceFactory_1.globalBasePath ? RestServiceFactory_1.globalBasePath : '';
    }
    setDomain(DOMAIN) {
        this.DOMAIN = DOMAIN;
    }
    setBasePath(basePath) {
        this.basePath = basePath;
    }
    get(uri, identifier = this.IDENTIFIER) {
        if (this.map.has(uri + identifier)) {
            return this.map.get(uri + identifier);
        }
        if (!/^https?\:\/\//.test(uri)) {
            const newBasePath = this.getNewBasePath();
            const basePathURI = lodash.isEmpty(newBasePath)
                ? uri
                : newBasePath + (/^\//.test(uri) ? uri : `/${uri}`);
            uri = this.shouldAppendDomain(uri) ? `${this.DOMAIN}/${uri}` : basePathURI;
        }
        const restClient = new RestClient(this.httpClient, uri, identifier);
        this.map.set(uri + identifier, restClient);
        return restClient;
    }
    shouldAppendDomain(uri) {
        return !!this.DOMAIN && !/^\//.test(uri);
    }
    getNewBasePath() {
        return this.basePath ? this.basePath : RestServiceFactory_1.globalBasePath;
    }
};
RestServiceFactory.globalBasePath = null;
RestServiceFactory.logService = new LogService();
RestServiceFactory.ctorParameters = () => [
    { type: HttpClient }
];
RestServiceFactory = RestServiceFactory_1 = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [HttpClient])
], RestServiceFactory);
export { RestServiceFactory };
