// @smart/utils/esm2015/services/authentication/authentication.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/*
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
import * as tslib_1 from "tslib";
import { Inject, Optional } from '@angular/core';
import * as lodash from 'lodash';
import { functionsUtils } from '../../utils';
import { IAuthenticationManagerService, IAuthenticationService, IModalService, ISettingsService, ISharedDataService, IStorageService } from '../../interfaces';
import { ITranslationsFetchService } from '../translations';
import { DEFAULT_AUTH_MAP, DEFAULT_AUTHENTICATION_ENTRY_POINT, DEFAULT_CREDENTIALS_MAP, EVENT_SERVICE, EVENTS } from '../../constants';
import { SSOAuthenticationHelper } from './sso-authentication.helper';
import { LoginDialogComponent } from '../../components/login-dialog';
let AuthenticationService = class AuthenticationService extends IAuthenticationService {
    constructor(translationsFetchService, modalService, sharedDataService, storageService, eventService, ssoAuthenticationHelper, settingsService, authenticationManager) {
        super();
        this.translationsFetchService = translationsFetchService;
        this.modalService = modalService;
        this.sharedDataService = sharedDataService;
        this.storageService = storageService;
        this.eventService = eventService;
        this.ssoAuthenticationHelper = ssoAuthenticationHelper;
        this.settingsService = settingsService;
        this.authenticationManager = authenticationManager;
    }
    filterEntryPoints(resource) {
        return this.sharedDataService.get('authenticationMap').then((authenticationMap) => {
            return functionsUtils
                .convertToArray(Object.assign({}, (authenticationMap || {}), DEFAULT_AUTH_MAP))
                .filter((entry) => new RegExp(entry.key, 'g').test(resource))
                .map((element) => element.value);
        });
    }
    isAuthEntryPoint(resource) {
        return this.sharedDataService.get('authenticationMap').then((authenticationMap) => {
            const authEntryPoints = functionsUtils
                .convertToArray(authenticationMap || {})
                .map((element) => element.value);
            return (authEntryPoints.indexOf(resource) > -1 ||
                resource === DEFAULT_AUTHENTICATION_ENTRY_POINT);
        });
    }
    authenticate(resource) {
        return this._findLoginData(resource).then((loginData) => {
            return this._launchAuth(loginData).then((modalFeedback) => {
                Promise.resolve(this.eventService.publish(EVENTS.AUTHORIZATION_SUCCESS, {
                    userHasChanged: modalFeedback.userHasChanged
                })).then(() => {
                    if (modalFeedback.userHasChanged) {
                        this.eventService.publish(EVENTS.USER_HAS_CHANGED);
                    }
                    /**
                     * We only need to reload when the user has changed and all authentication forms were closed.
                     * There can be many authentication forms if some modules use different (from default one) end points.
                     */
                    const reauthInProcess = lodash
                        .values(this.reauthInProgress)
                        .some((inProcess) => inProcess);
                    if (modalFeedback.userHasChanged &&
                        !reauthInProcess &&
                        this.authenticationManager &&
                        this.authenticationManager.onUserHasChanged) {
                        this.authenticationManager.onUserHasChanged();
                    }
                });
                this.reauthInProgress[loginData.authURI] = false;
            });
        });
    }
    logout() {
        // First, indicate the services that SmartEdit is logging out. This should give them the opportunity to clean up.
        // NOTE: This is not synchronous since some clean-up might be lengthy, and logging out should be fast.
        return this.eventService.publish(EVENTS.LOGOUT).then(() => {
            this.storageService.removeAllAuthTokens();
            if (this.ssoAuthenticationHelper.isAutoSSOMain()) {
                this.ssoAuthenticationHelper.logout();
            }
            else if (this.authenticationManager && this.authenticationManager.onLogout) {
                this.authenticationManager.onLogout();
            }
        });
    }
    isReAuthInProgress(entryPoint) {
        return Promise.resolve(this.reauthInProgress[entryPoint] === true);
    }
    setReAuthInProgress(entryPoint) {
        this.reauthInProgress[entryPoint] = true;
        return Promise.resolve();
    }
    isAuthenticated(url) {
        return this.filterEntryPoints(url).then((entryPoints) => {
            const authURI = entryPoints && entryPoints[0];
            return Promise.resolve(this.storageService.getAuthToken(authURI)).then((authToken) => {
                return !!authToken;
            });
        });
    }
    /*
     * will try determine first relevant authentication entry point from authenticationMap and retrieve potential client credentials to be added on top of user credentials
     */
    _findLoginData(resource) {
        return this.filterEntryPoints(resource).then((entryPoints) => {
            return Promise.resolve(this.sharedDataService.get('credentialsMap').then((credentialsMap) => {
                const map = Object.assign({}, (credentialsMap || {}), DEFAULT_CREDENTIALS_MAP);
                const authURI = entryPoints[0];
                return {
                    authURI,
                    clientCredentials: map[authURI]
                };
            }));
        });
    }
    _launchAuth(loginData) {
        return this.translationsFetchService
            .waitToBeReady()
            .then(() => Promise.all([
            this.storageService.isInitialized(),
            this.settingsService.getBoolean('smartedit.sso.enabled')
        ]))
            .then(([isFullScreen, ssoEnabled]) => {
            const modalRef = this.modalService.open({
                component: LoginDialogComponent,
                data: Object.assign({}, loginData, { isFullScreen,
                    ssoEnabled }),
                config: {
                    modalPanelClass: 'su-login-dialog-container',
                    hasBackdrop: false
                }
            });
            return new Promise((resolve, reject) => {
                modalRef.afterClosed.subscribe(resolve, reject);
            });
        });
    }
};
AuthenticationService = tslib_1.__decorate([
    tslib_1.__param(4, Inject(EVENT_SERVICE)),
    tslib_1.__param(7, Optional()),
    tslib_1.__metadata("design:paramtypes", [ITranslationsFetchService,
        IModalService,
        ISharedDataService,
        IStorageService, Object, SSOAuthenticationHelper,
        ISettingsService,
        IAuthenticationManagerService])
], AuthenticationService);
export { AuthenticationService };
