// @smart/utils/esm2015/services/browser/BrowserService.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export var SUPPORTED_BROWSERS;
(function (SUPPORTED_BROWSERS) {
    SUPPORTED_BROWSERS[SUPPORTED_BROWSERS["IE"] = 0] = "IE";
    SUPPORTED_BROWSERS[SUPPORTED_BROWSERS["CHROME"] = 1] = "CHROME";
    SUPPORTED_BROWSERS[SUPPORTED_BROWSERS["FIREFOX"] = 2] = "FIREFOX";
    SUPPORTED_BROWSERS[SUPPORTED_BROWSERS["EDGE"] = 3] = "EDGE";
    SUPPORTED_BROWSERS[SUPPORTED_BROWSERS["SAFARI"] = 4] = "SAFARI";
    SUPPORTED_BROWSERS[SUPPORTED_BROWSERS["UNKNOWN"] = 5] = "UNKNOWN";
})(SUPPORTED_BROWSERS || (SUPPORTED_BROWSERS = {}));
export class BrowserService {
    constructor() {
        /*
            It is always better to detect a browser via features. Unfortunately, it's becoming really hard to identify
            Safari, since newer versions do not match the previous ones. Thus, we have to rely on User Agent as the last
            option.
        */
        this._isSafari = () => {
            // return this.getCurrentBrowser() === SUPPORTED_BROWSERS.SAFARI;
            const userAgent = window.navigator.userAgent;
            const vendor = window.navigator.vendor;
            const testFeature = /constructor/i.test(function HTMLElementConstructor() {
                //
            }.toString());
            const testUserAgent = vendor && vendor.indexOf('Apple') > -1 && userAgent && !userAgent.match('CriOS');
            return testFeature || testUserAgent;
        };
        this.isIE = () => {
            return this.getCurrentBrowser() === SUPPORTED_BROWSERS.IE;
        };
        this.isFF = () => {
            return this.getCurrentBrowser() === SUPPORTED_BROWSERS.FIREFOX;
        };
        this.isSafari = () => {
            return this.getCurrentBrowser() === SUPPORTED_BROWSERS.SAFARI;
        };
    }
    getCurrentBrowser() {
        /* forbiddenNameSpaces window as any:false */
        const anyWindow = window;
        let browser = SUPPORTED_BROWSERS.UNKNOWN;
        if (typeof anyWindow.InstallTrigger !== 'undefined') {
            browser = SUPPORTED_BROWSERS.FIREFOX;
        }
        else if ( /*@cc_on!@*/false || !!document.documentMode) {
            browser = SUPPORTED_BROWSERS.IE;
        }
        else if (!!anyWindow.StyleMedia) {
            browser = SUPPORTED_BROWSERS.EDGE;
        }
        else if (!!anyWindow.chrome && !!anyWindow.chrome.webstore) {
            browser = SUPPORTED_BROWSERS.CHROME;
        }
        else if (this._isSafari()) {
            browser = SUPPORTED_BROWSERS.SAFARI;
        }
        return browser;
    }
    getBrowserLocale() {
        const locale = window.navigator.language.split('-');
        return locale.length === 1 ? locale[0] : locale[0] + '_' + locale[1].toUpperCase();
    }
}
