// @smart/utils/esm2015/services/browser/FingerPrintingService.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { stringUtils } from '../../utils';
/** @internal */
export class FingerPrintingService {
    constructor() {
        this._fingerprint = stringUtils.encode(this.getUserAgent() +
            this.getPlugins() +
            this.hasJava() +
            this.hasFlash() +
            this.hasLocalStorage() +
            this.hasSessionStorage() +
            this.hasCookie() +
            this.getTimeZone() +
            this.getLanguage() +
            this.getSystemLanguage() +
            this.hasCanvas());
    }
    /**
     * Get unique browser fingerprint information encoded in Base64.
     */
    getFingerprint() {
        return this._fingerprint;
    }
    getUserAgent() {
        return navigator.userAgent;
    }
    getPlugins() {
        const plugins = [];
        for (let i = 0, l = navigator.plugins.length; i < l; i++) {
            if (navigator.plugins[i]) {
                plugins.push(navigator.plugins[i].name);
            }
        }
        return plugins.join(',');
    }
    hasJava() {
        return navigator.javaEnabled();
    }
    hasFlash() {
        return !!navigator.plugins.namedItem('Shockwave Flash');
    }
    hasLocalStorage() {
        try {
            const hasLs = 'test-has-ls';
            localStorage.setItem(hasLs, hasLs);
            localStorage.removeItem(hasLs);
            return true;
        }
        catch (exception) {
            return false;
        }
    }
    hasSessionStorage() {
        try {
            const hasSs = 'test-has-ss';
            sessionStorage.setItem(hasSs, hasSs);
            sessionStorage.removeItem(hasSs);
            return true;
        }
        catch (exception) {
            return false;
        }
    }
    hasCookie() {
        return navigator.cookieEnabled;
    }
    getTimeZone() {
        return String(String(new Date()).split('(')[1]).split(')')[0];
    }
    getLanguage() {
        return navigator.language;
    }
    getSystemLanguage() {
        return window.navigator.language;
    }
    hasCanvas() {
        try {
            const elem = document.createElement('canvas');
            return !!(elem.getContext && elem.getContext('2d'));
        }
        catch (e) {
            return false;
        }
    }
}
