// @fundamental-ngx/core/esm5/lib/side-navigation/side-navigation-item/side-navigation-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, ViewEncapsulation } from '@angular/core';
import { SideNavigationLinkDirective } from '../side-navigation-link/side-navigation-link.directive';
import { SideNavigationSublistDirective } from '../side-navigation-sublist/side-navigation-sublist.directive';
/**
 * The component that represents a navigation group.
 * ```html
 * <fd-side-nav>
 *    <fd-side-nav-group>
 *        <h1 fd-side-nav-title>Group Name</h1>
 *          <div fd-side-nav-list>
 *             <fd-side-nav-item>
 *                <a fd-side-nav-link [attr.href]="'#'">Link Item</a>
 *             </fd-side-nav-item>
 *          </div>>
 *    </fd-side-nav-group>
 * </fd-side-nav>
 * ```
 */
var SideNavigationItemComponent = /** @class */ (function () {
    function SideNavigationItemComponent() {
    }
    /**
     * @return {?}
     */
    SideNavigationItemComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.linkElement && this.subListElement) {
            /** After view content check if there is flag with opened true */
            this.subListElement.subListIsOpenChange(this.linkElement.sublistIsOpen);
            this.subListOpenChanged$ = this.linkElement.onSubListOpenChange.subscribe((/**
             * @param {?} isOpen
             * @return {?}
             */
            function (isOpen) {
                _this.subListElement.subListIsOpenChange(isOpen);
            }));
        }
    };
    /**
     * @return {?}
     */
    SideNavigationItemComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.subListOpenChanged$) {
            this.subListOpenChanged$.unsubscribe();
        }
    };
    SideNavigationItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-side-nav-item',
                    template: "<li class=\"fd-side-nav__item\">\n  <ng-content select=\"[fd-side-nav-link]\"></ng-content>\n  <ng-content select=\"[fd-side-nav-sublist]\"></ng-content>\n</li>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    SideNavigationItemComponent.propDecorators = {
        linkElement: [{ type: ContentChild, args: [SideNavigationLinkDirective, { static: false },] }],
        subListElement: [{ type: ContentChild, args: [SideNavigationSublistDirective, { static: false },] }]
    };
    return SideNavigationItemComponent;
}());
export { SideNavigationItemComponent };
if (false) {
    /** @type {?} */
    SideNavigationItemComponent.prototype.linkElement;
    /** @type {?} */
    SideNavigationItemComponent.prototype.subListElement;
    /**
     * \@Hidden
     * @type {?}
     */
    SideNavigationItemComponent.prototype.subListOpenChanged$;
}
