// @smart/utils/esm5/modules/form-builder/models/abstract-form.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
var AbstractForm = /** @class */ (function () {
    function AbstractForm(component, control, inputs, persist, parent) {
        if (parent === void 0) { parent = null; }
        this.component = component;
        this.control = control;
        this.inputs = inputs;
        this.persist = persist;
        this.parent = parent;
        /**
         * Listen onto any property changes triggered by 'setInput' method.
         *
         * @type {Observable<InputPropertyChange>}
         */
        this.inputChanges = this.inputs.changes;
    }
    /**
     * Sets property and marks components for changes if property is decorated
     * with FormProperty.
     *
     * @param {keyof T} key
     * @param {T[keyof T]} value
     */
    AbstractForm.prototype.setInput = function (key, value) {
        this.inputs.set(key, value);
    };
    /**
     * Retrieves a property.
     *
     * @param {keyof T} key
     * @returns {T[keyof T] | undefined}
     */
    AbstractForm.prototype.getInput = function (key) {
        return this.inputs.get(key);
    };
    Object.defineProperty(AbstractForm.prototype, "root", {
        /**
         * Returns root form.
         */
        get: function () {
            if (this.parent === null) {
                return this;
            }
            var root = this.parent;
            while (root.parent) {
                root = root.parent;
            }
            return root;
        },
        enumerable: true,
        configurable: true
    });
    return AbstractForm;
}());
export { AbstractForm };
