// @smart/utils/esm5/modules/form-builder/models/validator-parameters.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { isUndefined, omitBy } from 'lodash';
/**
 * ValidatorParameters holds data to the synchronous and
 * asynchronous validators configuration for a FormField.
 */
var ValidatorParameters = /** @class */ (function () {
    function ValidatorParameters(validators, asyncValidators) {
        if (validators === void 0) { validators = {}; }
        if (asyncValidators === void 0) { asyncValidators = {}; }
        this.validators = {};
        this.asyncValidators = {};
        this.validators = this._omitUndefinedValues(validators);
        this.asyncValidators = this._omitUndefinedValues(asyncValidators);
    }
    /**
     * Determines if synchronous validator exists.
     *
     * @param name The name of the synchronous validator.
     * @returns A boolean if it has that parameter.
     */
    ValidatorParameters.prototype.has = function (name) {
        return this.validators.hasOwnProperty(name);
    };
    /**
     * Returns parameters of the synchronous validator.
     *
     * @param name The name of the synchronous validator.
     * @returns The param of the validator.
     */
    ValidatorParameters.prototype.getParam = function (name) {
        if (!this.has(name)) {
            return null;
        }
        return this.validators[name];
    };
    /**
     * Determines if asynchronous validator exists.
     *
     * @param name The name of the asynchronous validator.
     * @returns A boolean if it has that parameter.
     */
    ValidatorParameters.prototype.hasAsync = function (name) {
        return this.asyncValidators.hasOwnProperty(name);
    };
    /**
     * Returns parameters of the asynchronous validator.
     *
     * @param name The name of the asynchronous validator.
     * @returns The param of the validator.
     */
    ValidatorParameters.prototype.getAsyncParam = function (name) {
        if (!this.hasAsync(name)) {
            return null;
        }
        return this.asyncValidators[name];
    };
    /**
     * @internal
     * Returns a object with all those keys that have
     * undefined values.
     */
    ValidatorParameters.prototype._omitUndefinedValues = function (object) {
        return omitBy(object, isUndefined);
    };
    return ValidatorParameters;
}());
export { ValidatorParameters };
