// @smart/utils/esm2015/services/alert/alert.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
export class Alert {
    constructor(_alertConf, ALERT_CONFIG_DEFAULTS, fundamentalAlertService, translateService) {
        this._alertConf = _alertConf;
        this.fundamentalAlertService = fundamentalAlertService;
        this.translateService = translateService;
        this._displayed = false;
        lodash.defaultsDeep(this._alertConf, lodash.cloneDeep(ALERT_CONFIG_DEFAULTS));
    }
    get alertConf() {
        return this._alertConf;
    }
    get message() {
        return this._alertConf.message;
    }
    get type() {
        return this._alertConf.type;
    }
    /**
     * Displays the alert to the user.
     */
    show() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.isDisplayed()) {
                return;
            }
            if (this._alertConf.message) {
                this._alertConf.message = yield this.translateService
                    .get(this._alertConf.message, this._alertConf.messagePlaceholders)
                    .toPromise();
            }
            const content = typeof this._alertConf.message !== 'undefined'
                ? this._alertConf.message
                : this._alertConf.component || '';
            this._alertRef = this.fundamentalAlertService.open(content, this._alertConf);
            this._alertRef.afterDismissed.subscribe(() => (this._displayed = false));
            this._displayed = true;
        });
    }
    /**
     * Hides the alert if it is currently being displayed to the user.
     */
    hide() {
        if (!this.isDisplayed()) {
            return;
        }
        this._alertRef.dismiss();
    }
    isDisplayed() {
        return this._displayed;
    }
}
