// @smart/utils/esm2015/services/alert/alert.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { Injectable } from '@angular/core';
import { AlertFactory } from './alert-factory';
/**
 * @ngdoc service
 * @name @smartutils.services:AlertService
 */
let AlertService = class AlertService {
    constructor(alertFactory) {
        this.alertFactory = alertFactory;
    }
    /**
     * @ngdoc method
     * @name @smartutils.services:AlertService#showAlert
     * @methodOf @smartutils.services:AlertService
     * @param {string | Object} alertConf The alert's configuration {@link @smartutils.interfaces:IAlertConfig IAlertConfig}
     * @description
     * Displays an alert to the user. <br />
     * Convenience method to create an alert and call.show() on it immediately.
     */
    showAlert(alertConf) {
        const alert = this.alertFactory.createAlert(alertConf);
        alert.show();
    }
    /**
     * @ngdoc method
     * @name @smartutils.services:AlertService#showInfo
     * @methodOf @smartutils.services:AlertService
     * @param {string | Object} alertConf The alert's configuration {@link @smartutils.interfaces:IAlertConfig IAlertConfig}
     * @description
     * Displays an alert to the user. <br />
     * Convenience method to create an alert and call.show() on it immediately.
     */
    showInfo(alertConf) {
        const alert = this.alertFactory.createInfo(alertConf);
        alert.show();
    }
    /**
     * @ngdoc method
     * @name @smartutils.services:AlertService#showDanger
     * @methodOf @smartutils.services:AlertService
     * @param {string | Object} alertConf The alert's configuration {@link @smartutils.interfaces:IAlertConfig IAlertConfig}
     * @description
     * Displays an alert to the user. <br />
     * Convenience method to create an alert and call.show() on it immediately.
     */
    showDanger(alertConf) {
        const alert = this.alertFactory.createDanger(alertConf);
        alert.show();
    }
    /**
     * @ngdoc method
     * @name @smartutils.services:AlertService#showWarning
     * @methodOf @smartutils.services:AlertService
     * @param {string | Object} alertConf The alert's configuration {@link @smartutils.interfaces:IAlertConfig IAlertConfig}
     * @description
     * Displays an alert to the user. <br />
     * Convenience method to create an alert and call.show() on it immediately.
     */
    showWarning(alertConf) {
        const alert = this.alertFactory.createWarning(alertConf);
        alert.show();
    }
    /**
     * @ngdoc method
     * @name @smartutils.services:AlertService#showSuccess
     * @methodOf @smartutils.services:AlertService
     * @param {string | Object} alertConf The alert's configuration {@link @smartutils.interfaces:IAlertConfig IAlertConfig}
     * @description
     * Displays an alert to the user. <br />
     * Convenience method to create an alert and call.show() on it immediately.
     */
    showSuccess(alertConf) {
        const alert = this.alertFactory.createSuccess(alertConf);
        alert.show();
    }
};
AlertService.ctorParameters = () => [
    { type: AlertFactory }
];
AlertService = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [AlertFactory])
], AlertService);
export { AlertService };
