// @fundamental-ngx/core/esm5/lib/form/form-message/form-message.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewEncapsulation } from '@angular/core';
/**
 * Form message. Intended to be displayed under a form control for validation purposes.
 */
var FormMessageComponent = /** @class */ (function () {
    function FormMessageComponent() {
        /**
         * Type of the message. Can be `text`, `help`, `error` and `warning`.
         */
        this.type = '';
    }
    FormMessageComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-form-message',
                    template: "<span class=\"fd-form-message\" [ngClass]=\"[type ? ' fd-form-message--' + type : '']\">\n  <ng-content></ng-content>\n</span>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["@charset \"UTF-8\";/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-message(--help, --error, --warning)\n*/.fd-form-message{color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:8px 0;margin:8px 0 0;border:0;clear:both;display:block;font-size:12px;line-height:1.33333;font-weight:400;color:#51555a;color:var(--fd-color-text-2,#51555a);position:relative}.fd-form-message:after,.fd-form-message:before{box-sizing:inherit;font-size:inherit}.fd-form-message:before{width:18px;height:18px;font-style:normal;font-size:1.3em;font-family:SAP-icons;position:absolute;left:0;color:#51555a;color:var(--fd-color-text-2,#51555a);text-align:center;line-height:18px}.fd-form-message--error,.fd-form-message--help,.fd-form-message--warning{padding-left:24px}.fd-form-message--error:before,.fd-form-message--help:before,.fd-form-message--warning:before{left:0;top:6px}.fd-form-message--help:before{color:#32363a;color:var(--fd-color-text-1,#32363a);content:\"\uE1C3\"}.fd-form-message--warning{color:#e9730c;color:var(--fd-color-alert,#e9730c)}.fd-form-message--warning:before{color:#e9730c;color:var(--fd-color-alert,#e9730c);content:\"\uE053\"}.fd-form-message--error{color:#b00;color:var(--fd-color-negative,#b00)}.fd-form-message--error:before{content:\"\uE1EC\";color:#b00;color:var(--fd-color-negative,#b00)}"]
                }] }
    ];
    FormMessageComponent.propDecorators = {
        type: [{ type: Input }]
    };
    return FormMessageComponent;
}());
export { FormMessageComponent };
if (false) {
    /**
     * Type of the message. Can be `text`, `help`, `error` and `warning`.
     * @type {?}
     */
    FormMessageComponent.prototype.type;
}
