// @smart/utils/esm2015/modules/form-builder/models/abstract-form.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export class AbstractForm {
    constructor(component, control, inputs, persist, parent = null) {
        this.component = component;
        this.control = control;
        this.inputs = inputs;
        this.persist = persist;
        this.parent = parent;
        /**
         * Listen onto any property changes triggered by 'setInput' method.
         *
         * @type {Observable<InputPropertyChange>}
         */
        this.inputChanges = this.inputs.changes;
    }
    /**
     * Sets property and marks components for changes if property is decorated
     * with FormProperty.
     *
     * @param {keyof T} key
     * @param {T[keyof T]} value
     */
    setInput(key, value) {
        this.inputs.set(key, value);
    }
    /**
     * Retrieves a property.
     *
     * @param {keyof T} key
     * @returns {T[keyof T] | undefined}
     */
    getInput(key) {
        return this.inputs.get(key);
    }
    /**
     * Returns root form.
     */
    get root() {
        if (this.parent === null) {
            return this;
        }
        let root = this.parent;
        while (root.parent) {
            root = root.parent;
        }
        return root;
    }
}
