// @smart/utils/esm2015/modules/form-builder/models/validator-parameters.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { isUndefined, omitBy } from 'lodash';
/**
 * ValidatorParameters holds data to the synchronous and
 * asynchronous validators configuration for a FormField.
 */
export class ValidatorParameters {
    constructor(validators = {}, asyncValidators = {}) {
        this.validators = {};
        this.asyncValidators = {};
        this.validators = this._omitUndefinedValues(validators);
        this.asyncValidators = this._omitUndefinedValues(asyncValidators);
    }
    /**
     * Determines if synchronous validator exists.
     *
     * @param name The name of the synchronous validator.
     * @returns A boolean if it has that parameter.
     */
    has(name) {
        return this.validators.hasOwnProperty(name);
    }
    /**
     * Returns parameters of the synchronous validator.
     *
     * @param name The name of the synchronous validator.
     * @returns The param of the validator.
     */
    getParam(name) {
        if (!this.has(name)) {
            return null;
        }
        return this.validators[name];
    }
    /**
     * Determines if asynchronous validator exists.
     *
     * @param name The name of the asynchronous validator.
     * @returns A boolean if it has that parameter.
     */
    hasAsync(name) {
        return this.asyncValidators.hasOwnProperty(name);
    }
    /**
     * Returns parameters of the asynchronous validator.
     *
     * @param name The name of the asynchronous validator.
     * @returns The param of the validator.
     */
    getAsyncParam(name) {
        if (!this.hasAsync(name)) {
            return null;
        }
        return this.asyncValidators[name];
    }
    /**
     * @internal
     * Returns a object with all those keys that have
     * undefined values.
     */
    _omitUndefinedValues(object) {
        return omitBy(object, isUndefined);
    }
}
