// @smart/utils/esm2015/modules/form-builder/directives/form-renderer/form-renderer.directive.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { ComponentFactoryResolver, Directive, Input, OnDestroy, ViewContainerRef } from '@angular/core';
import { AbstractForm } from '../../models';
import { decorate } from './decorate';
let FormRendererDirective = class FormRendererDirective {
    constructor(componentFactoryResolver, viewContainerRef) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.viewContainerRef = viewContainerRef;
    }
    set form(form) {
        this._dispose();
        const componentFactory = this.componentFactoryResolver.resolveComponentFactory(form.component);
        // Create and decorate the component's inputs.
        const componentRef = this.viewContainerRef.createComponent(componentFactory);
        this._subscription = decorate(componentRef, form);
    }
    ngOnDestroy() {
        this._dispose();
    }
    /**
     * @internal
     * Clear all views and unsubscribe to streams.
     */
    _dispose() {
        this._subscription && this._subscription.unsubscribe();
        this.viewContainerRef.clear();
    }
};
FormRendererDirective.ctorParameters = () => [
    { type: ComponentFactoryResolver },
    { type: ViewContainerRef }
];
tslib_1.__decorate([
    Input('formRenderer'),
    tslib_1.__metadata("design:type", AbstractForm),
    tslib_1.__metadata("design:paramtypes", [AbstractForm])
], FormRendererDirective.prototype, "form", null);
FormRendererDirective = tslib_1.__decorate([
    Directive({
        selector: '[formRenderer]'
    }),
    tslib_1.__metadata("design:paramtypes", [ComponentFactoryResolver,
        ViewContainerRef])
], FormRendererDirective);
export { FormRendererDirective };
