// @fundamental-ngx/core/esm2015/lib/notification/notification-group/notification-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
export class NotificationGroupComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdNotificationGroupClass = true;
    }
}
NotificationGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-notification-group',
                template: "<ng-content select=\"fd-notification\"></ng-content>\n<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [""]
            }] }
];
NotificationGroupComponent.propDecorators = {
    fdNotificationGroupClass: [{ type: HostBinding, args: ['class.fd-notification--group',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationGroupComponent.prototype.fdNotificationGroupClass;
}
