// @smart/utils/esm5/services/interceptors/errors/permission-error.interceptor.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { Injectable } from '@angular/core';
import { IAlertService } from '../../../interfaces';
/**
 * @ngdoc service
 * @name permissionErrorInterceptorModule.service:permissionErrorInterceptor
 * @description
 * Used for HTTP error code 403. Displays the alert message for permission error.
 */
var PermissionErrorInterceptor = /** @class */ (function () {
    function PermissionErrorInterceptor(alertService) {
        this.alertService = alertService;
    }
    PermissionErrorInterceptor.prototype.predicate = function (request, response) {
        return response.status === 403;
    };
    PermissionErrorInterceptor.prototype.responseError = function (request, response) {
        var _this = this;
        if (response.error && response.error.errors) {
            response.error.errors
                .filter(function (error) {
                return error.type === 'TypePermissionError';
            })
                .forEach(function (error) {
                _this.alertService.showDanger({
                    message: error.message,
                    duration: 10000
                });
            });
        }
        return Promise.reject(response);
    };
    PermissionErrorInterceptor.ctorParameters = function () { return [
        { type: IAlertService }
    ]; };
    PermissionErrorInterceptor = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [IAlertService])
    ], PermissionErrorInterceptor);
    return PermissionErrorInterceptor;
}());
export { PermissionErrorInterceptor };
