// @smart/utils/esm2015/utils/http-utils.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { HttpErrorResponse, HttpParams, HttpResponse } from '@angular/common/http';
import { from, throwError, Observable } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { stringUtils } from './string-utils';
export class HttpUtils {
    isGET(request) {
        return request.method === 'GET';
    }
    isRequestOfAccept(request, accept) {
        return (!!request.headers &&
            !!request.headers.get('Accept') &&
            (request.headers.get('Accept') || '').includes(accept));
    }
    isResponseOfContentType(response, contentType) {
        return (!!response.headers &&
            !!response.headers.get('Content-type') &&
            (response.headers.get('Content-type') || '').indexOf(contentType) === 0);
    }
    isHTMLRequest(request, response) {
        return (this.isGET(request) &&
            (this.isRequestOfAccept(request, 'text/html') ||
                /.+\.html$/.test(request.url) ||
                /.+\.html\?/.test(request.url)));
    }
    isJSONRequest(request, response) {
        return (this.isGET(request) &&
            ((response && this.isResponseOfContentType(response, 'json')) ||
                /.+\.json$/.test(request.url)));
    }
    isJSRequest(request) {
        return this.isGET(request) && /.+\.js$/.test(request.url);
    }
    isCRUDRequest(request, response) {
        return (!this.isHTMLRequest(request, response) &&
            !this.isJSONRequest(request, response) &&
            !this.isJSRequest(request));
    }
    transformHttpParams(params, substitutionMap) {
        return new HttpParams({
            fromObject: JSON.parse(stringUtils.replaceAll(JSON.stringify(this.copyHttpParamsOrHeaders(params)), substitutionMap))
        });
    }
    copyHttpParamsOrHeaders(params) {
        const copy = {};
        params.keys().forEach((key) => {
            const values = params.getAll(key);
            if (values !== null) {
                copy[key] = values.length > 1 ? values : values[0];
            }
        });
        return copy;
    }
    buildHttpResponse(originalRequest, _statusAndPayload) {
        const statusAndPayloadPromise = Promise.resolve(_statusAndPayload);
        return from(statusAndPayloadPromise).pipe(switchMap((statusAndPayload) => {
            const status = statusAndPayload[0];
            const body = statusAndPayload[1];
            const requestClone = originalRequest.clone({
                body
            });
            lodash.merge(requestClone, { status });
            if (200 <= status && status < 300) {
                return new Observable((ob) => {
                    ob.next(new HttpResponse(requestClone));
                });
            }
            else {
                return throwError(new HttpErrorResponse(lodash.merge(requestClone, { error: body })));
            }
        }));
    }
}
export const httpUtils = new HttpUtils();
