// @smart/utils/esm2015/utils/promise-utils.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
var PromiseUtils_1;
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { Injectable } from '@angular/core';
import { LogService } from '../services/log.service';
/**
 * @ngdoc service
 * @name @smartutils.services:PromiseUtils
 *
 * @description
 * utility service around ES6 Promises.
 */
let PromiseUtils = PromiseUtils_1 = class PromiseUtils {
    /**
     * @ngdoc service
     * @name @smartutils.services:PromiseUtils
     *
     * @description
     * utility service around ES6 Promises.
     */
    constructor() {
        this.WAIT_TIMEOUT = 4;
        this.FAILURE_TIMEOUT = 2000;
        this.handlePromiseRejections = (promise) => {
            const oldThen = promise.then;
            const defaultFailureCallback = this.defaultFailureCallback;
            promise.then = function (successCallback, _failureCallback) {
                const failureCallback = _failureCallback ? _failureCallback : defaultFailureCallback;
                return oldThen.call(this, successCallback, failureCallback);
            };
            return promise;
        };
        this.defaultFailureCallback = (error) => {
            if (undefined !== error && null !== error && 'canceled' !== error) {
                if (lodash.isPlainObject(error)) {
                    if (!this.isAjaxError(error)) {
                        PromiseUtils_1.logService.error(`exception caught in promise: ${JSON.stringify(error)}`);
                    }
                }
                else if (!lodash.isBoolean(error)) {
                    PromiseUtils_1.logService.error(error);
                }
            }
            PromiseUtils_1.logService.error(`defaultFailureCallback:`, error);
            return Promise.reject(error);
        };
    }
    toPromise(method, context) {
        return function () {
            try {
                return Promise.resolve(method.apply(context, Array.prototype.slice.call(arguments)));
            }
            catch (e) {
                PromiseUtils_1.logService.error('execution of a method that was turned into a promise failed');
                PromiseUtils_1.logService.error(e);
                return Promise.reject(e);
            }
        };
    }
    promise(executor) {
        return this.handlePromiseRejections(new Promise(executor));
    }
    defer() {
        let pResolve;
        let pReject;
        const deferred = {
            promise: this.promise((_resolve, _reject) => {
                pResolve = _resolve;
                pReject = _reject;
            }),
            resolve(value) {
                pResolve(value);
            },
            reject(reason) {
                pReject(reason);
            }
        };
        return deferred;
    }
    sleep(ms) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => setTimeout(resolve, ms));
        });
    }
    isAjaxError(error) {
        return error.hasOwnProperty('headers');
    }
    waitOnCondition(condition, callback, errorMessage, elapsedTime = 0) {
        setTimeout(() => {
            if (condition()) {
                callback();
            }
            else if (elapsedTime < this.FAILURE_TIMEOUT) {
                this.waitOnCondition(condition, callback, errorMessage, elapsedTime + this.WAIT_TIMEOUT);
            }
            else {
                throw new Error(`PromiseUtils: ${errorMessage}`);
            }
        }, this.WAIT_TIMEOUT);
    }
    resolveToCallbackWhenCondition(condition, callback, errorMessage) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                this.waitOnCondition(condition, () => resolve(callback()), errorMessage ? errorMessage : 'condition for promise resolution was never met');
            });
        });
    }
};
PromiseUtils.logService = new LogService();
PromiseUtils = PromiseUtils_1 = tslib_1.__decorate([
    Injectable()
], PromiseUtils);
export { PromiseUtils };
export const promiseUtils = new PromiseUtils();
