// @fundamental-ngx/core/esm2015/lib/input-group/input-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, forwardRef, ViewEncapsulation, ContentChild, TemplateRef, } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { InputGroupAddOnDirective, InputGroupInputDirective } from './input-group-directives';
/**
 * The component that represents an input group.
 * The input group includes form inputs with add-ons that allow the user to better understand the information being entered.
 *
 * ```html
 * <fd-input-group [placement]="'after'" [addOnText]="'$'" [placeholder]="'Amount'">
 * </fd-input-group>
 * ```
 */
export class InputGroupComponent {
    constructor() {
        /**
         * The placement of the add-on.
         * Options include *before* and *after*
         */
        this.placement = 'after';
        /**
         * Whether the input group is in compact mode.
         */
        this.compact = false;
        /**
         * Whether Button should be focusable
         */
        this.buttonFocusable = true;
        /**
         * Button options.  Options include 'emphasized' and 'light'. Leave empty for default.'
         */
        this.buttonOptions = 'light';
        /**
         * Event emitted when the add-on button is clicked.
         */
        this.addOnButtonClicked = new EventEmitter();
        /**
         * @hidden
         */
        this.onChange = (/**
         * @return {?}
         */
        () => { });
        /**
         * @hidden
         */
        this.onTouched = (/**
         * @return {?}
         */
        () => { });
    }
    /**
     * Get the value of the text input.
     * @return {?}
     */
    get inputText() {
        return this.inputTextValue;
    }
    /**
     * Set the value of the text input.
     * @param {?} value
     * @return {?}
     */
    set inputText(value) {
        this.inputTextValue = value;
        this.onChange(value);
        this.onTouched();
    }
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    writeValue(value) {
        this.inputTextValue = value;
    }
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    registerOnChange(fn) {
        this.onChange = fn;
    }
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    /**
     * @hidden
     * @param {?} $event
     * @return {?}
     */
    buttonClicked($event) {
        this.addOnButtonClicked.emit($event);
    }
}
InputGroupComponent.decorators = [
    { type: Component, args: [{
                selector: ' fd-input-group',
                template: "<div class=\"fd-input-group\"\n     [ngClass]=\"{\n        'fd-input-group--after': placement === 'after',\n        'fd-input-group--before': placement === 'before',\n        'fd-input-group--inline': inline === true,\n        'fd-input-group--compact' : compact}\">\n\n    <ng-container *ngIf=\"placement !== 'before'\">\n        <ng-container *ngTemplateOutlet=\"input\"></ng-container>\n    </ng-container>\n\n\n    <ng-content *ngIf=\"addOnElement\" select=\"[fd-input-group-addon]\"></ng-content>\n\n    <ng-container *ngIf=\"!addOnElement\">\n\n        <span *ngIf=\"button\"\n              fd-input-group-addon\n              [button]=\"true\"\n              [placement]=\"placement\">\n            <button [disabled]=\"disabled\"\n                    fd-button\n                    [compact]=\"compact\"\n                    [attr.tabindex]=\"buttonFocusable ? 0 : -1\"\n                    [fdType]=\"buttonType\"\n                    [options]=\"buttonOptions\"\n                    [glyph]=\"glyph\"\n                    (click)=\"buttonClicked($event)\">\n                <ng-container *ngIf=\"!glyph\">{{addOnText}}</ng-container>\n            </button>\n        </span>\n\n        <span *ngIf=\"!button\"\n              fd-input-group-addon\n              [placement]=\"placement\">\n            <ng-container *ngIf=\"!glyph\">{{addOnText}}</ng-container>\n            <span *ngIf=\"glyph\" [ngClass]=\"'sap-icon--' + glyph\"></span>\n    </span>\n    </ng-container>\n\n    <ng-container *ngIf=\"placement === 'before'\">\n        <ng-container *ngTemplateOutlet=\"input\"></ng-container>\n    </ng-container>\n</div>\n\n<ng-template #input>\n    <ng-content *ngIf=\"inputElement\" select=\"[fd-input-group-input]\"></ng-content>\n    <ng-container *ngIf=\"inputTemplate\">\n        <ng-container *ngTemplateOutlet=\"inputTemplate\"></ng-container>\n    </ng-container>\n    <input [(ngModel)]=\"inputText\"\n           [disabled]=\"disabled\"\n           type=\"text\"\n           *ngIf=\"!inputElement && !inputTemplate\"\n           placeholder=\"{{placeholder}}\"\n           class=\"fd-input fd-input-group__default-input\"\n           [ngClass]=\"{'fd-input--compact' : compact}\">\n</ng-template>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => InputGroupComponent)),
                        multi: true
                    }
                ],
                encapsulation: ViewEncapsulation.None,
                styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-input-group+(--inline)\n    .fd-input-group__addon+(--button)\n*/.fd-input-group{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:flex}.fd-input-group:after,.fd-input-group:before{box-sizing:inherit;font-size:inherit}.fd-input-group>:first-child{border-top-right-radius:0;border-bottom-right-radius:0}.fd-input-group>:last-child{border-top-left-radius:0;border-bottom-left-radius:0}.fd-input-group>:not(:first-child):not(:last-child){border-radius:0}.fd-input-group>[dir=rtl]:first-child,[dir=rtl] .fd-input-group>:first-child{border-radius:0 4px 4px 0}.fd-input-group>[dir=rtl]:last-child,[dir=rtl] .fd-input-group>:last-child{border-radius:4px 0 0 4px}.fd-input-group .fd-input-group__input{width:100%;z-index:1}.fd-input-group .fd-input-group__input:nth-child(n+1){margin:0}.fd-input-group--inline{display:inline-flex;vertical-align:bottom}.fd-input-group__addon{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 12px;margin:0;border:1px solid #89919a;flex:0 0 auto;z-index:auto;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);border-color:var(--fd-forms-border-color,#89919a);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa);justify-content:center;display:flex;flex-direction:column;border-radius:4px}.fd-input-group__addon:after,.fd-input-group__addon:before{box-sizing:inherit;font-size:inherit}.fd-input-group__addon--readonly{opacity:.4}.fd-input-group__addon:first-child{margin-right:-1px}.fd-input-group__addon:last-child{margin-left:-1px}.fd-input-group__addon[dir=rtl]:first-child,[dir=rtl] .fd-input-group__addon:first-child{margin-left:-1px;margin-right:0}.fd-input-group__addon[dir=rtl]:last-child,[dir=rtl] .fd-input-group__addon:last-child{margin-right:-1px;margin-left:0}.fd-input-group__addon--button{border:none;padding:0}.fd-input-group__addon--button .fd-input-group__button{height:100%;max-height:100%;border-color:var(--fd-forms-border-color,#89919a);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa);border:1px solid #89919a}.fd-input-group__addon--button .fd-input-group__button:nth-child(n+1){margin:0}.fd-input-group__addon--button:first-child>*{margin-right:-1px;border-top-right-radius:0;border-bottom-right-radius:0}.fd-input-group__addon--button:first-child>:first-child:not(:last-child){border-bottom-left-radius:0}.fd-input-group__addon--button:first-child>:last-child:not(:first-child){border-top-left-radius:0;margin-top:-1px}.fd-input-group__addon--button:last-child>*{margin-left:-1px;border-top-left-radius:0;border-bottom-left-radius:0}.fd-input-group__addon--button:last-child>:first-child:not(:last-child){border-bottom-right-radius:0}.fd-input-group__addon--button:last-child>:last-child:not(:first-child){border-top-right-radius:0;margin-top:-1px}.fd-input-group__addon--button[dir=rtl]:first-child>*,[dir=rtl] .fd-input-group__addon--button:first-child>*{margin-left:-1px;margin-right:0;border-radius:0 4px 4px 0}.fd-input-group__addon--button[dir=rtl]:first-child>:first-child:not(:last-child),[dir=rtl] .fd-input-group__addon--button:first-child>:first-child:not(:last-child){border-bottom-right-radius:0}.fd-input-group__addon--button[dir=rtl]:first-child>:last-child:not(:first-child),[dir=rtl] .fd-input-group__addon--button:first-child>:last-child:not(:first-child){border-top-right-radius:0;margin-top:-1px}.fd-input-group__addon--button[dir=rtl]:last-child>*,[dir=rtl] .fd-input-group__addon--button:last-child>*{margin-right:-1px;margin-left:0;border-radius:4px 0 0 4px}.fd-input-group__addon--button[dir=rtl]:last-child>:first-child:not(:last-child),[dir=rtl] .fd-input-group__addon--button:last-child>:first-child:not(:last-child){border-bottom-left-radius:0}.fd-input-group__addon--button[dir=rtl]:last-child>:last-child:not(:first-child),[dir=rtl] .fd-input-group__addon--button:last-child>:last-child:not(:first-child){border-top-left-radius:0;margin-top:-1px}/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-input\n*/.fd-input{color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;-webkit-font-smoothing:antialiased;-webkit-box-sizing:border-box;padding:0 12px;margin:0;border:1px solid #89919a;font-size:14px;line-height:1.42857;font-weight:400;appearance:none;-webkit-appearance:textfield;-moz-appearance:textfield;box-sizing:border-box;outline:0;border-color:var(--fd-forms-border-color,#89919a);background-color:#fff;background-color:var(--fd-forms-background-color,#fff);border-radius:4px;transition:border-color 125ms;height:36px;height:var(--fd-forms-height,36px);width:100%}.fd-input:after,.fd-input:before{box-sizing:inherit;font-size:inherit}.fd-input .is-hover,.fd-input:hover{border-color:#085caf;border-color:var(--fd-color-action-hover,#085caf)}.fd-input.is-focus,.fd-input:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;border-color:#085caf;border-color:var(--fd-color-action-focus,#085caf)}.fd-input.is-invalid{border-color:#b00;border-color:var(--fd-color-negative,#b00);border-width:2px}.fd-input.is-valid{border-color:#0a7e3e;border-color:var(--fd-color-positive,#0a7e3e);border-width:2px}.fd-input.is-alert,.fd-input.is-warning{border-color:#e9730c;border-color:var(--fd-color-alert,#e9730c);border-width:2px}.fd-input.is-disabled,.fd-input:disabled,.fd-input[aria-disabled=true]{cursor:not-allowed;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa)}.fd-input.is-readonly,.fd-input[readonly]{color:#32363a;color:var(--fd-forms-color,#32363a);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);border-width:0 0 1px;border-radius:0}.fd-input::-webkit-input-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::-moz-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input:-ms-input-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::-ms-input-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::-webkit-input-placeholder:after,.fd-input::-webkit-input-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input::-moz-placeholder:after,.fd-input::-moz-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input:-ms-input-placeholder:after,.fd-input:-ms-input-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input::-ms-input-placeholder:after,.fd-input::-ms-input-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input::placeholder:after,.fd-input::placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input--compact{height:28px;height:var(--fd-forms-height-compact,28px)}.fd-input--no-number-spinner{-moz-appearance:textfield}.fd-input--no-number-spinner::-webkit-inner-spin-button,.fd-input--no-number-spinner::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-textarea\n*/.fd-textarea{color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;-webkit-font-smoothing:antialiased;-webkit-box-sizing:border-box;padding:12px 12px 0;margin:0;border:1px solid #89919a;font-size:14px;line-height:1.42857;font-weight:400;appearance:none;-webkit-appearance:textfield;-moz-appearance:textfield;box-sizing:border-box;outline:0;border-color:var(--fd-forms-border-color,#89919a);background-color:#fff;background-color:var(--fd-forms-background-color,#fff);border-radius:4px;transition:border-color 125ms;height:72px;width:100%}.fd-textarea:after,.fd-textarea:before{box-sizing:inherit;font-size:inherit}.fd-textarea .is-hover,.fd-textarea:hover{border-color:#085caf;border-color:var(--fd-color-action-hover,#085caf)}.fd-textarea.is-focus,.fd-textarea:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;border-color:#085caf;border-color:var(--fd-color-action-focus,#085caf)}.fd-textarea.is-invalid{border-color:#b00;border-color:var(--fd-color-negative,#b00);border-width:2px}.fd-textarea.is-valid{border-color:#0a7e3e;border-color:var(--fd-color-positive,#0a7e3e);border-width:2px}.fd-textarea.is-alert,.fd-textarea.is-warning{border-color:#e9730c;border-color:var(--fd-color-alert,#e9730c);border-width:2px}.fd-textarea.is-disabled,.fd-textarea:disabled,.fd-textarea[aria-disabled=true]{cursor:not-allowed;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa)}.fd-textarea.is-readonly,.fd-textarea[readonly]{color:#32363a;color:var(--fd-forms-color,#32363a);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);border-width:0 0 1px;border-radius:0}.fd-textarea--compact{height:56px;padding-top:12px}fd-input-group .fd-input-group__addon--button.fd-input-group__addon--textarea>*{max-height:100%;height:100%}.fd-input-group .fd-input-group__addon--button>*{border-color:var(--fd-forms-border-color,#89919a);height:100%;max-height:100%}"]
            }] }
];
InputGroupComponent.propDecorators = {
    inputElement: [{ type: ContentChild, args: [InputGroupInputDirective, { static: false },] }],
    addOnElement: [{ type: ContentChild, args: [InputGroupAddOnDirective, { static: false },] }],
    inputTemplate: [{ type: Input }],
    placement: [{ type: Input }],
    compact: [{ type: Input }],
    inline: [{ type: Input }],
    placeholder: [{ type: Input }],
    addOnText: [{ type: Input }],
    buttonFocusable: [{ type: Input }],
    buttonType: [{ type: Input }],
    buttonOptions: [{ type: Input }],
    glyph: [{ type: Input }],
    button: [{ type: Input }],
    disabled: [{ type: Input }],
    addOnButtonClicked: [{ type: Output }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    InputGroupComponent.prototype.inputElement;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupComponent.prototype.addOnElement;
    /**
     * Input template
     * @type {?}
     */
    InputGroupComponent.prototype.inputTemplate;
    /**
     * The placement of the add-on.
     * Options include *before* and *after*
     * @type {?}
     */
    InputGroupComponent.prototype.placement;
    /**
     * Whether the input group is in compact mode.
     * @type {?}
     */
    InputGroupComponent.prototype.compact;
    /**
     * Whether the input group is inline.
     * @type {?}
     */
    InputGroupComponent.prototype.inline;
    /**
     * Placeholder for the input group.
     * @type {?}
     */
    InputGroupComponent.prototype.placeholder;
    /**
     * The text for the add-on.
     * @type {?}
     */
    InputGroupComponent.prototype.addOnText;
    /**
     * Whether Button should be focusable
     * @type {?}
     */
    InputGroupComponent.prototype.buttonFocusable;
    /**
     * The type of the button. Types include 'standard', 'positive', 'medium', and 'negative'.
     * Leave empty for default (Action button).'
     * @type {?}
     */
    InputGroupComponent.prototype.buttonType;
    /**
     * Button options.  Options include 'emphasized' and 'light'. Leave empty for default.'
     * @type {?}
     */
    InputGroupComponent.prototype.buttonOptions;
    /**
     * The icon value for the add-on.
     * @type {?}
     */
    InputGroupComponent.prototype.glyph;
    /**
     * Whether the icon add-on or the text add-on is a button.
     * @type {?}
     */
    InputGroupComponent.prototype.button;
    /**
     * Whether the input group is disabled.
     * @type {?}
     */
    InputGroupComponent.prototype.disabled;
    /**
     * Event emitted when the add-on button is clicked.
     * @type {?}
     */
    InputGroupComponent.prototype.addOnButtonClicked;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupComponent.prototype.inputTextValue;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupComponent.prototype.onChange;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupComponent.prototype.onTouched;
}
