// @fundamental-ngx/core/esm5/lib/localizator-editor/localization-editor-item/localization-editor-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, Input, TemplateRef, ViewEncapsulation } from '@angular/core';
import { LocalizationEditorInputDirective, LocalizationEditorLabel, LocalizationEditorTextareaDirective } from '../localization-editor.directives';
/**
 *  Component that represents field with add-on.
 *  ```html
 *  <fd-localization-editor-item [label]="'EN'">
 *      <input fd-localization-editor-input type="text" placeholder="EN">
 *  </fd-localization-editor-item>
 *  ```
 */
var LocalizationEditorItemComponent = /** @class */ (function () {
    function LocalizationEditorItemComponent() {
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    LocalizationEditorItemComponent.prototype.ngOnInit = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.refreshChildInput();
    };
    /**
     * @return {?}
     */
    LocalizationEditorItemComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.refreshChildInput();
    };
    /**
     * @return {?}
     */
    LocalizationEditorItemComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        if (this.textarea) {
            this.type = 'textarea';
        }
    };
    /**
     * @private
     * @return {?}
     */
    LocalizationEditorItemComponent.prototype.refreshChildInput = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.input) {
            this.input.compact = this.compact;
        }
        if (this.textarea) {
            this.textarea.compact = this.compact;
        }
    };
    LocalizationEditorItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-localization-editor-item',
                    template: "<fd-input-group [compact]=\"compact\" [inputTemplate]=\"input\">\n    <span fd-input-group-addon [type]=\"type\">\n        <ng-container *ngIf=\"labelTemplate\">\n            <ng-container [fd-localization-editor-load-label]=\"labelTemplate\"></ng-container>\n        </ng-container>\n        <ng-container *ngIf=\"!labelTemplate\">\n            {{label}}\n        </ng-container>\n    </span>\n</fd-input-group>\n<ng-template #input>\n    <ng-content select=\"[fd-localization-editor-input]\"></ng-content>\n    <ng-content select=\"[fd-localization-editor-textarea]\"></ng-content>\n</ng-template>\n<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None
                }] }
    ];
    LocalizationEditorItemComponent.propDecorators = {
        label: [{ type: Input }],
        compact: [{ type: Input }],
        input: [{ type: ContentChild, args: [LocalizationEditorInputDirective, { static: false },] }],
        textarea: [{ type: ContentChild, args: [LocalizationEditorTextareaDirective, { static: false },] }],
        labelTemplate: [{ type: ContentChild, args: [LocalizationEditorLabel, { read: TemplateRef, static: false },] }]
    };
    return LocalizationEditorItemComponent;
}());
export { LocalizationEditorItemComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorItemComponent.prototype.type;
    /**
     * The text for the add-on on the right side.
     * @type {?}
     */
    LocalizationEditorItemComponent.prototype.label;
    /**
     * Whether to apply compact mode to to field.
     * @type {?}
     */
    LocalizationEditorItemComponent.prototype.compact;
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorItemComponent.prototype.input;
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorItemComponent.prototype.textarea;
    /**
     * @hidden
     * @type {?}
     */
    LocalizationEditorItemComponent.prototype.labelTemplate;
}
