// @smart/utils/esm5/services/modal/modal.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { Injectable } from '@angular/core';
import { ModalConfig as FundamentalModalConfig, ModalRef as FundamentalModalRef, ModalService as FundamentalModalService } from '@fundamental-ngx/core';
import { FundamentalModalTemplateComponent } from '../../components';
var ModalService = /** @class */ (function () {
    function ModalService(fundamentalModalService) {
        this.fundamentalModalService = fundamentalModalService;
    }
    ModalService.prototype.open = function (options) {
        var templateConfig = options.templateConfig;
        return !!templateConfig
            ? this.openWithTemplate(options)
            : this.openStandalone(options);
    };
    ModalService.prototype.openStandalone = function (options) {
        var _a = options.config, config = _a === void 0 ? {} : _a, component = options.component, data = options.data;
        return this.fundamentalModalService.open(component, tslib_1.__assign({}, config, { data: data }));
    };
    ModalService.prototype.openWithTemplate = function (options) {
        var _a = options.config, config = _a === void 0 ? {} : _a, _b = options.templateConfig, templateConfig = _b === void 0 ? {} : _b, component = options.component, data = options.data;
        var settings = tslib_1.__assign({}, config, { data: {
                templateConfig: templateConfig,
                component: component,
                modalData: data
            } });
        return this.fundamentalModalService.open(FundamentalModalTemplateComponent, settings);
    };
    ModalService.ctorParameters = function () { return [
        { type: FundamentalModalService }
    ]; };
    ModalService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [FundamentalModalService])
    ], ModalService);
    return ModalService;
}());
export { ModalService };
