// @smart/utils/esm2015/services/interceptors/http-interceptor.module.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
var HttpInterceptorModule_1;
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { APP_BOOTSTRAP_LISTENER, NgModule } from '@angular/core';
import { defaultRetryStrategyFactory, exponentialRetryStrategyFactory, linearRetryStrategyFactory, DefaultRetryStrategy, ExponentialRetry, ExponentialRetryStrategy, LinearRetry, LinearRetryStrategy, SimpleRetry } from './errors';
import { BackendInterceptor, HttpBackendService } from './backend';
import { FlawInjectionInterceptorModule } from './flaws';
import { HttpAuthInterceptor } from './http-auth.interceptor';
import { HttpErrorInterceptorService } from './http-error-interceptor.service';
import { HttpErrorInterceptor } from './http-error.interceptor';
const ɵ0 = defaultRetryStrategyFactory, ɵ1 = exponentialRetryStrategyFactory, ɵ2 = linearRetryStrategyFactory;
/**
 * @ngdoc overview
 * @name httpInterceptorModule
 *
 * @description
 * This module provides the functionality to add custom HTTP error interceptors.
 * Interceptors are used to execute code each time an HTTP request fails.
 *
 */
let HttpInterceptorModule = HttpInterceptorModule_1 = class HttpInterceptorModule {
    static forRoot(...HttpErrorInterceptorClasses) {
        return {
            ngModule: HttpInterceptorModule_1,
            providers: [
                ...HttpErrorInterceptorClasses,
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: HttpAuthInterceptor,
                    multi: true
                },
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: HttpErrorInterceptor,
                    multi: true
                },
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: BackendInterceptor,
                    multi: true
                },
                {
                    provide: APP_BOOTSTRAP_LISTENER,
                    useFactory(httpErrorInterceptorService) {
                        httpErrorInterceptorService.addInterceptors(HttpErrorInterceptorClasses);
                        return (component) => {
                            // an initializer useFactory must return a function
                        };
                    },
                    deps: [HttpErrorInterceptorService],
                    multi: true
                }
            ]
        };
    }
};
HttpInterceptorModule = HttpInterceptorModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [FlawInjectionInterceptorModule],
        providers: [
            SimpleRetry,
            LinearRetry,
            ExponentialRetry,
            {
                provide: DefaultRetryStrategy,
                useFactory: ɵ0,
                deps: [SimpleRetry]
            },
            {
                provide: ExponentialRetryStrategy,
                useFactory: ɵ1,
                deps: [ExponentialRetry]
            },
            {
                provide: LinearRetryStrategy,
                useFactory: ɵ2,
                deps: [SimpleRetry]
            },
            HttpErrorInterceptorService,
            HttpBackendService
        ]
    })
], HttpInterceptorModule);
export { HttpInterceptorModule };
export { ɵ0, ɵ1, ɵ2 };
