// @smart/utils/esm5/modules/form-builder/decorators.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/* tslint:disable:max-classes-per-file */
import * as tslib_1 from "tslib";
/**
 * @internal
 * Internal property on the constructor used for adding decorator metadata
 * so that it can be later picked up after component compilation.
 */
export var FORM_PROP = '_form_prop_';
/**
 * Base data PropDecorator.
 * @internal
 */
var PropDecorator = /** @class */ (function () {
    function PropDecorator(property) {
        this.property = property;
    }
    return PropDecorator;
}());
export { PropDecorator };
/**
 * @internal
 */
var InputPropDecorator = /** @class */ (function (_super) {
    tslib_1.__extends(InputPropDecorator, _super);
    function InputPropDecorator(property, alias) {
        var _this = _super.call(this, property) || this;
        _this.alias = alias ? alias : _this.property;
        return _this;
    }
    return InputPropDecorator;
}(PropDecorator));
export { InputPropDecorator };
/**
 * @internal
 */
var FormPropDecorator = /** @class */ (function (_super) {
    tslib_1.__extends(FormPropDecorator, _super);
    function FormPropDecorator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FormPropDecorator;
}(PropDecorator));
export { FormPropDecorator };
/**
 * @internal
 */
var ControlPropDecorator = /** @class */ (function (_super) {
    tslib_1.__extends(ControlPropDecorator, _super);
    function ControlPropDecorator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ControlPropDecorator;
}(PropDecorator));
export { ControlPropDecorator };
/**
 * Used for tagging dynamic inputs and adding them to the FORM_PROP property
 * of the target constructor.
 */
function makePropertyDecorator(factory) {
    return function (target, key) {
        var ctor = target.constructor;
        if (!ctor[FORM_PROP]) {
            ctor[FORM_PROP] = [];
        }
        ctor[FORM_PROP].push(factory(key));
    };
}
/**
 * Injects the AbstractForm for the dynamic form component.
 */
export function DynamicForm() {
    return makePropertyDecorator(function (key) { return new FormPropDecorator(key); });
}
/**
 * Injects the AbstractFormControl of the AbstractForm for the dynamic form component.
 */
export function DynamicFormControl() {
    return makePropertyDecorator(function (key) { return new ControlPropDecorator(key); });
}
/**
 * Injects a property of the AbstractForm for the dynamic form component.
 * Inputs are assigned from the FormSchema's 'inputs' property.
 * NOTE:
 * Property values are only available ngOnInit or onDynamicInputChange.
 * @param alias Use this alias to target a property of the AbstractForm. Defaults
 * to the assigned class property.
 * @example
 * <pre>
 *     @Component({ ... })
 *     export class DynamicFormComponent {
 *         @DynamicInput()
 *         property: string
 *     }
 * <pre>
 */
export var DynamicInput = function (alias) {
    if (alias === void 0) { alias = null; }
    return makePropertyDecorator(function (key) { return new InputPropDecorator(key, alias); });
};
