// @smart/utils/esm5/modules/form-builder/form-builder.module.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
// File named form-builder.module for now cuz of build blocking me to use Module suffix.
import * as tslib_1 from "tslib";
import { ANALYZE_FOR_ENTRY_COMPONENTS, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
// Components
import { FormListerComponent } from './components/form-lister/form-lister.component';
import { FormRendererDirective } from './directives/form-renderer/form-renderer.directive';
// Services
import { ASYNC_VALIDATOR_MAP, COMPONENT_MAP, VALIDATOR_MAP } from './services/injection-tokens';
var FormBuilderModule = /** @class */ (function () {
    function FormBuilderModule() {
    }
    FormBuilderModule_1 = FormBuilderModule;
    FormBuilderModule.forRoot = function (option) {
        return {
            ngModule: FormBuilderModule_1,
            providers: [
                {
                    provide: ANALYZE_FOR_ENTRY_COMPONENTS,
                    useValue: [option.types],
                    multi: true
                },
                {
                    provide: COMPONENT_MAP,
                    useValue: option.types
                },
                {
                    provide: VALIDATOR_MAP,
                    useValue: option.validators
                },
                {
                    provide: ASYNC_VALIDATOR_MAP,
                    useValue: option.asyncValidators
                }
            ]
        };
    };
    var FormBuilderModule_1;
    FormBuilderModule = FormBuilderModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [CommonModule, ReactiveFormsModule],
            declarations: [FormRendererDirective, FormListerComponent],
            entryComponents: [FormListerComponent],
            exports: [FormRendererDirective]
        })
    ], FormBuilderModule);
    return FormBuilderModule;
}());
export { FormBuilderModule };
