// @smart/utils/esm2015/components/language-dropdown/language-dropdown-helper.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export var LanguageSortStrategy;
(function (LanguageSortStrategy) {
    LanguageSortStrategy["Default"] = "default";
    LanguageSortStrategy["None"] = "none";
})(LanguageSortStrategy || (LanguageSortStrategy = {}));
// @dynamic
export class LanguageDropdownHelper {
    static findSelectLanguageWithIsoCodePredicate(isoCode) {
        return (item) => item.value.isoCode === isoCode;
    }
    /**
     * Finds the language with a specified isoCode.
     *
     * @param {string} isoCode
     * @param {IToolingLanguage[]} languages
     * @returns {IToolingLanguage}
     */
    static findLanguageWithIsoCode(isoCode, languages) {
        return languages.find((language) => language.isoCode === isoCode);
    }
    /**
     * Returns an ordered language array by name and sets the selected language at the beginning.
     *
     * @param {IToolingLanguage} selectedLanguage
     * @param {IToolingLanguage[]} languages
     * @param {ILanguageSortConfig} config
     * @returns {IToolingLanguage[]}
     */
    static order(selectedLanguage, languages, config = { strategy: LanguageSortStrategy.Default }) {
        switch (config.strategy) {
            case LanguageSortStrategy.Default:
                const orderedLanguages = languages
                    .filter((language) => language.isoCode !== selectedLanguage.isoCode)
                    .sort((a, b) => {
                    return a.isoCode.localeCompare(b.isoCode);
                });
                return [selectedLanguage, ...orderedLanguages];
            case LanguageSortStrategy.None:
                return languages;
        }
    }
}
