// @smart/utils/esm2015/services/log.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["log"] = 0] = "log";
    LogLevel[LogLevel["debug"] = 1] = "debug";
    LogLevel[LogLevel["info"] = 2] = "info";
    LogLevel[LogLevel["warn"] = 3] = "warn";
    LogLevel[LogLevel["error"] = 4] = "error";
})(LogLevel || (LogLevel = {}));
export class LogService {
    constructor() {
        this.logLevel = LogLevel.info;
    }
    log(...msg) {
        this._log(LogLevel.log, msg);
    }
    debug(...msg) {
        this._log(LogLevel.debug, msg);
    }
    info(...msg) {
        this._log(LogLevel.info, msg);
    }
    warn(...msg) {
        this._log(LogLevel.warn, msg);
    }
    error(...msg) {
        this._log(LogLevel.error, msg);
    }
    setLogLevel(logLevel) {
        this.logLevel = logLevel;
    }
    _log(requestLevel, msg) {
        if (requestLevel >= this.logLevel) {
            const method = LogLevel[requestLevel];
            if (this._console() && this._console()[method]) {
                this._console()[method](...msg);
            }
        }
    }
    _console() {
        return console;
    }
}
