// @smart/utils/esm5/services/interceptors/errors/retries/retryPolicies/simple-retry.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * @ngdoc object
 * @name @smartutils.object:SIMPLE_RETRY_DEFAULT_SETTING
 *
 * @description
 * The setting object to be used as default values for retry.
 */
export var SIMPLE_RETRY_DEFAULT_SETTING = {
    MAX_ATTEMPT: 5,
    MIN_BACKOFF: 0,
    RETRY_INTERVAL: 500
};
/**
 * @ngdoc service
 * @name @smartutils.services:simpleRetry
 * @description
 * When used by a retry strategy, this service could provide a simple fixed delay time to be used by the strategy before the next request is sent. The service also provides functionality to check if it is possible to perform a next retry.
 */
var SimpleRetry = /** @class */ (function () {
    function SimpleRetry() {
    }
    SimpleRetry.prototype.calculateNextDelay = function (retryInterval, minBackoff) {
        minBackoff = minBackoff || SIMPLE_RETRY_DEFAULT_SETTING.MIN_BACKOFF;
        retryInterval = retryInterval || SIMPLE_RETRY_DEFAULT_SETTING.RETRY_INTERVAL;
        var waveShield = minBackoff + Math.random();
        return retryInterval + waveShield;
    };
    SimpleRetry.prototype.canRetry = function (attemptCount, _maxAttempt) {
        var maxAttempt = _maxAttempt || SIMPLE_RETRY_DEFAULT_SETTING.MAX_ATTEMPT;
        return attemptCount <= maxAttempt;
    };
    return SimpleRetry;
}());
export { SimpleRetry };
