// @smart/utils/esm5/services/interceptors/errors/retries/operation-context.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { Injectable } from '@angular/core';
/**
 * @ngdoc service
 * @name @smartutils.services:OperationContextService
 * @description
 * This service provides the functionality to register a url with its associated operation contexts and also finds operation context given an url.
 */
var OperationContextService = /** @class */ (function () {
    function OperationContextService() {
        this.store = [];
    }
    /**
     * @ngdoc method
     * @name @smartutils.services:OperationContextService#register
     * @methodOf @smartutils.services:OperationContextService
     *
     * @description
     * Register a new url with it's associated operationContext.
     *
     * @param {String} url The url that is associated to the operation context.
     * @param {String} operationContext The operation context name that is associated to the given url.
     *
     * @return {Object} operationContextService The operationContextService service
     */
    OperationContextService.prototype.register = function (url, operationContext) {
        if (typeof url !== 'string' || lodash.isEmpty(url)) {
            throw new Error('operationContextService.register error: url is invalid');
        }
        if (typeof operationContext !== 'string' || lodash.isEmpty(operationContext)) {
            throw new Error('operationContextService.register error: operationContext is invalid');
        }
        var regexIndex = this.store.findIndex(function (store) {
            return store.urlRegex.test(url) === true && store.operationContext === operationContext;
        });
        if (regexIndex !== -1) {
            return this;
        }
        var urlRegex = new RegExp(url.replace(/\/:[^\/]*/g, '/.*'));
        this.store.push({
            urlRegex: urlRegex,
            operationContext: operationContext
        });
        return this;
    };
    /**
     * @ngdoc method
     * @name @smartutils.services:OperationContextService#findOperationContext
     * @methodOf @smartutils.services:OperationContextService
     *
     * @description
     * Find the first matching operation context for the given url.
     *
     * @param {String} url The request url.
     *
     * @return {String} operationContext
     */
    OperationContextService.prototype.findOperationContext = function (url) {
        var regexIndex = this.store.findIndex(function (store) { return store.urlRegex.test(url) === true; });
        return ~regexIndex ? this.store[regexIndex].operationContext : null;
    };
    OperationContextService = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [])
    ], OperationContextService);
    return OperationContextService;
}());
export { OperationContextService };
