// @smart/utils/esm2015/directives/list-keyboard-control/list-item-keyboard-control.directive.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, OnInit, Renderer2 } from '@angular/core';
/**
 * Directive for marking list item for 'ListKeyboardControlDirective' to allow for navigating with keyboard.
 */
let ListItemKeyboardControlDirective = class ListItemKeyboardControlDirective {
    constructor(hostElement, renderer) {
        this.hostElement = hostElement;
        this.renderer = renderer;
        /** @internal */
        this.activeClassName = 'is-active';
    }
    /** @internal */
    ngOnInit() {
        this.setTabIndex();
    }
    getElement() {
        return this.hostElement.nativeElement;
    }
    setActive() {
        const elm = this.getElement();
        this.renderer.addClass(elm, this.activeClassName);
        elm.scrollIntoView({ block: 'nearest' });
    }
    setInactive() {
        this.renderer.removeClass(this.getElement(), this.activeClassName);
    }
    /** @internal */
    setTabIndex() {
        this.renderer.setAttribute(this.getElement(), 'tabindex', '-1');
    }
};
ListItemKeyboardControlDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
ListItemKeyboardControlDirective = tslib_1.__decorate([
    Directive({
        selector: '[suListItemKeyboardControl]'
    }),
    tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
], ListItemKeyboardControlDirective);
export { ListItemKeyboardControlDirective };
