// @smart/utils/esm5/services/interceptors/backend/backend.interceptor.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { Injectable } from '@angular/core';
import { take } from 'rxjs/operators';
import { HttpUtils, UrlUtils } from '../../../utils';
import { LogService } from '../../../services/log.service';
import { HttpBackendService } from './http-backend.service';
/*
 * This is the place where the entries through HttpBackenService invocations are being used.
 * All outbound http requests are funneled through here, when a match from HttpBackenService entries
 * is found, the request is intercepted and the specified mock is returned with especified status code.
 * If no match is found, the http request is effectively sent over the wire
 */
var BackendInterceptor = /** @class */ (function () {
    function BackendInterceptor(httpBackendService, httpUtils, urlUtils, logService) {
        this.httpBackendService = httpBackendService;
        this.httpUtils = httpUtils;
        this.urlUtils = urlUtils;
        this.logService = logService;
    }
    BackendInterceptor.prototype.intercept = function (request, next) {
        var backendMockRespond = this.httpBackendService.findMatchingMock(request);
        if (!backendMockRespond) {
            return next.handle(request);
        }
        var response;
        if (typeof backendMockRespond === 'object') {
            response = [200, lodash.cloneDeep(backendMockRespond)];
        }
        else {
            // if (typeof backendMockRespond === 'function')
            var data = null;
            if (request.method === 'GET') {
                data = decodeURIComponent(this.urlUtils.getQueryString(this.httpUtils.copyHttpParamsOrHeaders(request.params)));
            }
            else if (request.headers.get('Content-Type') === 'application/x-www-form-urlencoded') {
                data = request.body; // it is a query string
            }
            else if (request.method === 'POST' || request.method === 'PUT') {
                data = JSON.stringify(request.body);
            }
            var headers = this.httpUtils.copyHttpParamsOrHeaders(request.headers);
            response = backendMockRespond(request.method, decodeURIComponent(request.urlWithParams), data, headers);
        }
        this.logService.debug("backend " + status + " response for " + request.url + ": ");
        return this.httpUtils.buildHttpResponse(request, response).pipe(take(1));
    };
    BackendInterceptor.ctorParameters = function () { return [
        { type: HttpBackendService },
        { type: HttpUtils },
        { type: UrlUtils },
        { type: LogService }
    ]; };
    BackendInterceptor = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [HttpBackendService,
            HttpUtils,
            UrlUtils,
            LogService])
    ], BackendInterceptor);
    return BackendInterceptor;
}());
export { BackendInterceptor };
