// @smart/utils/esm2015/services/modal/modal.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { Injectable } from '@angular/core';
import { ModalConfig as FundamentalModalConfig, ModalRef as FundamentalModalRef, ModalService as FundamentalModalService } from '@fundamental-ngx/core';
import { FundamentalModalTemplateComponent } from '../../components';
let ModalService = class ModalService {
    constructor(fundamentalModalService) {
        this.fundamentalModalService = fundamentalModalService;
    }
    open(options) {
        const { templateConfig } = options;
        return !!templateConfig
            ? this.openWithTemplate(options)
            : this.openStandalone(options);
    }
    openStandalone(options) {
        const { config = {}, component, data } = options;
        return this.fundamentalModalService.open(component, Object.assign({}, config, { data }));
    }
    openWithTemplate(options) {
        const { config = {}, templateConfig = {}, component, data } = options;
        const settings = Object.assign({}, config, { data: {
                templateConfig,
                component,
                modalData: data
            } });
        return this.fundamentalModalService.open(FundamentalModalTemplateComponent, settings);
    }
};
ModalService.ctorParameters = () => [
    { type: FundamentalModalService }
];
ModalService = tslib_1.__decorate([
    Injectable(),
    tslib_1.__metadata("design:paramtypes", [FundamentalModalService])
], ModalService);
export { ModalService };
