// @fundamental-ngx/core/esm5/lib/notification/notification-footer/notification-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
var NotificationFooterComponent = /** @class */ (function () {
    function NotificationFooterComponent() {
        /**
         * @hidden
         */
        this.fdNotificationFooterClass = true;
    }
    NotificationFooterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-notification-footer',
                    template: "<ng-content></ng-content>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: [""]
                }] }
    ];
    NotificationFooterComponent.propDecorators = {
        fdNotificationFooterClass: [{ type: HostBinding, args: ['class.fd-notification__footer',] }]
    };
    return NotificationFooterComponent;
}());
export { NotificationFooterComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    NotificationFooterComponent.prototype.fdNotificationFooterClass;
}
