// @smart/utils/esm2015/components/modal-template/modal-template.component.js
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/*
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
import { Component, ViewEncapsulation } from '@angular/core';
import { map } from 'rxjs/operators';
import { FundamentalModalManagerService } from '../../services/modal/modal-manager.service';
let FundamentalModalTemplateComponent = class FundamentalModalTemplateComponent {
    constructor(modalManager) {
        this.modalManager = modalManager;
        this.component = this.modalManager.getComponent();
        this.title = this.modalManager.getTitle();
        this.titleSuffix = this.modalManager.getTitleSuffix();
        this.buttons = this.modalManager.getButtons();
        this.isDismissButtonVisible = this.modalManager.getIsDismissButtonVisible();
    }
    ngOnInit() {
        this.modalManager.init();
    }
    onButtonClicked(button) {
        this.modalManager.onButtonClicked(button);
    }
    dismiss() {
        this.modalManager.dismiss();
    }
    hasButtons() {
        return this.buttons.pipe(map((buttons) => buttons && !!buttons.length));
    }
};
FundamentalModalTemplateComponent.ctorParameters = () => [
    { type: FundamentalModalManagerService }
];
FundamentalModalTemplateComponent = tslib_1.__decorate([
    Component({
        selector: 'fundamental-modal-template',
        encapsulation: ViewEncapsulation.None,
        providers: [FundamentalModalManagerService],
        template: `
        <fd-modal-header>
            <h1 fd-modal-title id="fd-modal-title-{{ (title | async) || '' }}">
                {{ (title | async) || '' | translate }}
                {{ (titleSuffix | async) || '' | translate }}
            </h1>
            <button
                fd-modal-close-btn
                *ngIf="isDismissButtonVisible | async"
                (click)="dismiss()"
            ></button>
        </fd-modal-header>
        <fd-modal-body>
            <ng-container *ngComponentOutlet="component | async"></ng-container>
        </fd-modal-body>
        <fd-modal-footer *ngIf="hasButtons() | async">
            <button
                *ngFor="let button of buttons | async"
                [disabled]="button.disabledFn ? button.disabledFn() : button.disabled"
                [options]="button.style"
                [attr.id]="button.id"
                (click)="onButtonClicked(button)"
                fd-button
            >
                {{ button.label | translate }}
            </button>
        </fd-modal-footer>
    `,
        styles: [`
            .fd-modal__title {
                min-height: 20px;
            }
        `]
    }),
    tslib_1.__metadata("design:paramtypes", [FundamentalModalManagerService])
], FundamentalModalTemplateComponent);
export { FundamentalModalTemplateComponent };
