// @fundamental-ngx/core/esm5/lib/form/form-label/form-label.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Label to be linked to a form control.
 *
 * ```html
 * <label fd-form-label for="input-id">Label Text</label>
 * <input fd-form-control type="text" id="input-id" />
 * ```
 */
var FormLabelComponent = /** @class */ (function () {
    function FormLabelComponent() {
        /**
         * @hidden
         */
        this.fdFormLabelClass = true;
    }
    FormLabelComponent.decorators = [
        { type: Component, args: [{
                    // TODO to be discussed
                    // tslint:disable-next-line:component-selector
                    selector: '[fd-form-label]',
                    template: "<ng-content></ng-content>",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-label\n*/.fd-form-label{font-size:14px;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 8px;border:0;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);cursor:pointer;line-height:16px;display:block}.fd-form-label:after,.fd-form-label:before{box-sizing:inherit;font-size:inherit}.fd-form-label.is-disabled{opacity:.4}.fd-form-label:focus-within{outline-color:var(--fd-color-action-focus,#085caf);outline:#085caf dotted 1px}.fd-form-label .fd-checkbox,.fd-form-label>.fd-radio{margin-right:8px}"]
                }] }
    ];
    FormLabelComponent.propDecorators = {
        fdFormLabelClass: [{ type: HostBinding, args: ['class.fd-form-label',] }]
    };
    return FormLabelComponent;
}());
export { FormLabelComponent };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    FormLabelComponent.prototype.fdFormLabelClass;
}
