// @smart/utils/esm5/services/interceptors/flaws/flaw-injection-interceptor.Module.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/* forbiddenNameSpaces useClass:false */
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { FlawInjectionInterceptor } from './flaw-injection.interceptor';
/** @internal */
var FlawInjectionInterceptorModule = /** @class */ (function () {
    function FlawInjectionInterceptorModule() {
    }
    FlawInjectionInterceptorModule = tslib_1.__decorate([
        NgModule({
            imports: [],
            providers: [
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: FlawInjectionInterceptor,
                    multi: true
                },
                {
                    provide: APP_INITIALIZER,
                    useFactory: function () {
                        FlawInjectionInterceptor.registerRequestFlaw({
                            test: function (request) { return /sites\/[\w-]+\//.test(request.url); },
                            mutate: function (request) {
                                return request.clone({
                                    url: request.url.replace(/sites\/([\w-]+)\//, 'sites/' + Math.random() + '/')
                                });
                            }
                        });
                        return function (component) {
                            // an initializer useFactory must return a function
                        };
                    },
                    multi: true
                }
            ]
        })
    ], FlawInjectionInterceptorModule);
    return FlawInjectionInterceptorModule;
}());
export { FlawInjectionInterceptorModule };
