// @smart/utils/esm5/services/alert/alert.module.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { NgModule } from '@angular/core';
import { AlertModule as FundamentalAlertModule } from '@fundamental-ngx/core';
import { AlertService } from './alert.service';
import { AlertFactory } from './alert-factory';
import { TranslationModule } from '../translations';
import { ALERT_CONFIG_DEFAULTS, ALERT_CONFIG_DEFAULTS_TOKEN } from './alert-token';
var ɵ0 = ALERT_CONFIG_DEFAULTS;
var AlertModule = /** @class */ (function () {
    function AlertModule() {
    }
    AlertModule = tslib_1.__decorate([
        NgModule({
            imports: [FundamentalAlertModule, TranslationModule],
            providers: [
                {
                    provide: ALERT_CONFIG_DEFAULTS_TOKEN,
                    useValue: ɵ0
                },
                AlertService,
                AlertFactory
            ]
        })
    ], AlertModule);
    return AlertModule;
}());
export { AlertModule };
export { ɵ0 };
