// @smart/utils/esm5/services/alert/alert.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
var Alert = /** @class */ (function () {
    function Alert(_alertConf, ALERT_CONFIG_DEFAULTS, fundamentalAlertService, translateService) {
        this._alertConf = _alertConf;
        this.fundamentalAlertService = fundamentalAlertService;
        this.translateService = translateService;
        this._displayed = false;
        lodash.defaultsDeep(this._alertConf, lodash.cloneDeep(ALERT_CONFIG_DEFAULTS));
    }
    Object.defineProperty(Alert.prototype, "alertConf", {
        get: function () {
            return this._alertConf;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Alert.prototype, "message", {
        get: function () {
            return this._alertConf.message;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Alert.prototype, "type", {
        get: function () {
            return this._alertConf.type;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Displays the alert to the user.
     */
    Alert.prototype.show = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _a, content;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.isDisplayed()) {
                            return [2 /*return*/];
                        }
                        if (!this._alertConf.message) return [3 /*break*/, 2];
                        _a = this._alertConf;
                        return [4 /*yield*/, this.translateService
                                .get(this._alertConf.message, this._alertConf.messagePlaceholders)
                                .toPromise()];
                    case 1:
                        _a.message = _b.sent();
                        _b.label = 2;
                    case 2:
                        content = typeof this._alertConf.message !== 'undefined'
                            ? this._alertConf.message
                            : this._alertConf.component || '';
                        this._alertRef = this.fundamentalAlertService.open(content, this._alertConf);
                        this._alertRef.afterDismissed.subscribe(function () { return (_this._displayed = false); });
                        this._displayed = true;
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Hides the alert if it is currently being displayed to the user.
     */
    Alert.prototype.hide = function () {
        if (!this.isDisplayed()) {
            return;
        }
        this._alertRef.dismiss();
    };
    Alert.prototype.isDisplayed = function () {
        return this._displayed;
    };
    return Alert;
}());
export { Alert };
