// @fundamental-ngx/core/esm2015/lib/form/form-group/form-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Represents a form group element.
 *
 * ```html
 * <fd-form-group>
 *     <div fd-form-item [isCheck]="true">
 *         <input fd-form-control type="radio" checked>
 *         <fd-form-label>Option One</fd-form-label>
 *     </div>
 * </fd-form-group>
 * ```
 */
export class FormGroupComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdFormGroupClass = true;
    }
}
FormGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-form-group',
                template: "<ng-content></ng-content>\n",
                encapsulation: ViewEncapsulation.None,
                styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-form-group\n*/.fd-form-group{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0 0 16px;border:0}.fd-form-group:after,.fd-form-group:before{box-sizing:inherit;font-size:inherit}.fd-form-group:last-child{margin-bottom:0}"]
            }] }
];
FormGroupComponent.propDecorators = {
    fdFormGroupClass: [{ type: HostBinding, args: ['class.fd-form-group',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    FormGroupComponent.prototype.fdFormGroupClass;
}
