// @fundamental-ngx/core/esm5/lib/input-group/input-group-number.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input, ViewEncapsulation } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/**
 * The component that represents an integer value input.
 * The value is increased or decreased using the spinner add-on.
 *
 * ```html
 * <fd-input-group-number [disabled]="false" [(ngModel)]="numberValue"></fd-input-group-number>
 * ```
 */
var InputGroupNumberComponent = /** @class */ (function () {
    function InputGroupNumberComponent() {
        /**
         * Aria label for the 'step up' button.
         */
        this.stepUpLabel = 'Step up';
        /**
         * Aria label for the 'step down' button.
         */
        this.stepDownLabel = 'Step down';
        /**
         * @hidden
         */
        this.onChange = (/**
         * @return {?}
         */
        function () { });
        /**
         * @hidden
         */
        this.onTouched = (/**
         * @return {?}
         */
        function () { });
    }
    Object.defineProperty(InputGroupNumberComponent.prototype, "inputText", {
        /** Get the value of the text input. */
        get: /**
         * Get the value of the text input.
         * @return {?}
         */
        function () {
            return this.inputTextValue;
        },
        /** Set the value of the text input. */
        set: /**
         * Set the value of the text input.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this.inputTextValue = value;
            this.onChange(value);
            this.onTouched();
        },
        enumerable: true,
        configurable: true
    });
    /** @hidden */
    /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    InputGroupNumberComponent.prototype.writeValue = /**
     * @hidden
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.inputTextValue = value;
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    InputGroupNumberComponent.prototype.registerOnChange = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onChange = fn;
    };
    /** @hidden */
    /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    InputGroupNumberComponent.prototype.registerOnTouched = /**
     * @hidden
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onTouched = fn;
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InputGroupNumberComponent.prototype.stepUpClicked = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.inputTextValue++;
        this.onChange(this.inputTextValue);
        this.onTouched();
    };
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InputGroupNumberComponent.prototype.stepDownClicked = /**
     * @hidden
     * @return {?}
     */
    function () {
        this.inputTextValue--;
        this.onChange(this.inputTextValue);
        this.onTouched();
    };
    InputGroupNumberComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-input-group-number',
                    template: "<div class=\"fd-input-group fd-input-group--after\">\n    <input type=\"number\"\n           class=\"fd-input fd-input-group__input\"\n           [disabled]=\"disabled\"\n           [(ngModel)]=\"inputText\"\n           placeholder=\"{{placeholder}}\"/>\n    <span fd-input-group-addon [button]=\"true\" [placement]=\"'after'\">\n        <button fd-button\n                [glyph]=\"'slim-arrow-up'\"\n                [fdType]=\"'half'\"\n                [options]=\"'light'\"\n                class=\"fd-input-group__button\"\n                [attr.aria-label]=\"stepUpLabel\"\n                (click)=\"stepUpClicked()\">\n        </button>\n        <button fd-button\n                [glyph]=\"'slim-arrow-down'\"\n                [fdType]=\"'half'\"\n                [options]=\"'light'\"\n                class=\"fd-input-group__button\"\n                [attr.aria-label]=\"stepDownLabel\"\n                (click)=\"stepDownClicked()\">\n        </button>\n    </span>\n</div>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return InputGroupNumberComponent; })),
                            multi: true
                        }
                    ],
                    encapsulation: ViewEncapsulation.None,
                    styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-input-group+(--inline)\n    .fd-input-group__addon+(--button)\n*/.fd-input-group{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:flex}.fd-input-group:after,.fd-input-group:before{box-sizing:inherit;font-size:inherit}.fd-input-group>:first-child{border-top-right-radius:0;border-bottom-right-radius:0}.fd-input-group>:last-child{border-top-left-radius:0;border-bottom-left-radius:0}.fd-input-group>:not(:first-child):not(:last-child){border-radius:0}.fd-input-group>[dir=rtl]:first-child,[dir=rtl] .fd-input-group>:first-child{border-radius:0 4px 4px 0}.fd-input-group>[dir=rtl]:last-child,[dir=rtl] .fd-input-group>:last-child{border-radius:4px 0 0 4px}.fd-input-group .fd-input-group__input{width:100%;z-index:1}.fd-input-group .fd-input-group__input:nth-child(n+1){margin:0}.fd-input-group--inline{display:inline-flex;vertical-align:bottom}.fd-input-group__addon{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0 12px;margin:0;border:1px solid #89919a;flex:0 0 auto;z-index:auto;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);border-color:var(--fd-forms-border-color,#89919a);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa);justify-content:center;display:flex;flex-direction:column;border-radius:4px}.fd-input-group__addon:after,.fd-input-group__addon:before{box-sizing:inherit;font-size:inherit}.fd-input-group__addon--readonly{opacity:.4}.fd-input-group__addon:first-child{margin-right:-1px}.fd-input-group__addon:last-child{margin-left:-1px}.fd-input-group__addon[dir=rtl]:first-child,[dir=rtl] .fd-input-group__addon:first-child{margin-left:-1px;margin-right:0}.fd-input-group__addon[dir=rtl]:last-child,[dir=rtl] .fd-input-group__addon:last-child{margin-right:-1px;margin-left:0}.fd-input-group__addon--button{border:none;padding:0}.fd-input-group__addon--button .fd-input-group__button{height:100%;max-height:100%;border-color:var(--fd-forms-border-color,#89919a);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa);border:1px solid #89919a}.fd-input-group__addon--button .fd-input-group__button:nth-child(n+1){margin:0}.fd-input-group__addon--button:first-child>*{margin-right:-1px;border-top-right-radius:0;border-bottom-right-radius:0}.fd-input-group__addon--button:first-child>:first-child:not(:last-child){border-bottom-left-radius:0}.fd-input-group__addon--button:first-child>:last-child:not(:first-child){border-top-left-radius:0;margin-top:-1px}.fd-input-group__addon--button:last-child>*{margin-left:-1px;border-top-left-radius:0;border-bottom-left-radius:0}.fd-input-group__addon--button:last-child>:first-child:not(:last-child){border-bottom-right-radius:0}.fd-input-group__addon--button:last-child>:last-child:not(:first-child){border-top-right-radius:0;margin-top:-1px}.fd-input-group__addon--button[dir=rtl]:first-child>*,[dir=rtl] .fd-input-group__addon--button:first-child>*{margin-left:-1px;margin-right:0;border-radius:0 4px 4px 0}.fd-input-group__addon--button[dir=rtl]:first-child>:first-child:not(:last-child),[dir=rtl] .fd-input-group__addon--button:first-child>:first-child:not(:last-child){border-bottom-right-radius:0}.fd-input-group__addon--button[dir=rtl]:first-child>:last-child:not(:first-child),[dir=rtl] .fd-input-group__addon--button:first-child>:last-child:not(:first-child){border-top-right-radius:0;margin-top:-1px}.fd-input-group__addon--button[dir=rtl]:last-child>*,[dir=rtl] .fd-input-group__addon--button:last-child>*{margin-right:-1px;margin-left:0;border-radius:4px 0 0 4px}.fd-input-group__addon--button[dir=rtl]:last-child>:first-child:not(:last-child),[dir=rtl] .fd-input-group__addon--button:last-child>:first-child:not(:last-child){border-bottom-left-radius:0}.fd-input-group__addon--button[dir=rtl]:last-child>:last-child:not(:first-child),[dir=rtl] .fd-input-group__addon--button:last-child>:last-child:not(:first-child){border-top-left-radius:0;margin-top:-1px}/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-input\n*/.fd-input{color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;-webkit-font-smoothing:antialiased;-webkit-box-sizing:border-box;padding:0 12px;margin:0;border:1px solid #89919a;font-size:14px;line-height:1.42857;font-weight:400;appearance:none;-webkit-appearance:textfield;-moz-appearance:textfield;box-sizing:border-box;outline:0;border-color:var(--fd-forms-border-color,#89919a);background-color:#fff;background-color:var(--fd-forms-background-color,#fff);border-radius:4px;transition:border-color 125ms;height:36px;height:var(--fd-forms-height,36px);width:100%}.fd-input:after,.fd-input:before{box-sizing:inherit;font-size:inherit}.fd-input .is-hover,.fd-input:hover{border-color:#085caf;border-color:var(--fd-color-action-hover,#085caf)}.fd-input.is-focus,.fd-input:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;border-color:#085caf;border-color:var(--fd-color-action-focus,#085caf)}.fd-input.is-invalid{border-color:#b00;border-color:var(--fd-color-negative,#b00);border-width:2px}.fd-input.is-valid{border-color:#0a7e3e;border-color:var(--fd-color-positive,#0a7e3e);border-width:2px}.fd-input.is-alert,.fd-input.is-warning{border-color:#e9730c;border-color:var(--fd-color-alert,#e9730c);border-width:2px}.fd-input.is-disabled,.fd-input:disabled,.fd-input[aria-disabled=true]{cursor:not-allowed;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa)}.fd-input.is-readonly,.fd-input[readonly]{color:#32363a;color:var(--fd-forms-color,#32363a);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);border-width:0 0 1px;border-radius:0}.fd-input::-webkit-input-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::-moz-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input:-ms-input-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::-ms-input-placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::placeholder{font-size:14px;line-height:1.42857;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;color:#8d9194;color:var(--fd-color-neutral-4,#8d9194)}.fd-input::-webkit-input-placeholder:after,.fd-input::-webkit-input-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input::-moz-placeholder:after,.fd-input::-moz-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input:-ms-input-placeholder:after,.fd-input:-ms-input-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input::-ms-input-placeholder:after,.fd-input::-ms-input-placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input::placeholder:after,.fd-input::placeholder:before{box-sizing:inherit;font-size:inherit}.fd-input--compact{height:28px;height:var(--fd-forms-height-compact,28px)}.fd-input--no-number-spinner{-moz-appearance:textfield}.fd-input--no-number-spinner::-webkit-inner-spin-button,.fd-input--no-number-spinner::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-textarea\n*/.fd-textarea{color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;-webkit-font-smoothing:antialiased;-webkit-box-sizing:border-box;padding:12px 12px 0;margin:0;border:1px solid #89919a;font-size:14px;line-height:1.42857;font-weight:400;appearance:none;-webkit-appearance:textfield;-moz-appearance:textfield;box-sizing:border-box;outline:0;border-color:var(--fd-forms-border-color,#89919a);background-color:#fff;background-color:var(--fd-forms-background-color,#fff);border-radius:4px;transition:border-color 125ms;height:72px;width:100%}.fd-textarea:after,.fd-textarea:before{box-sizing:inherit;font-size:inherit}.fd-textarea .is-hover,.fd-textarea:hover{border-color:#085caf;border-color:var(--fd-color-action-hover,#085caf)}.fd-textarea.is-focus,.fd-textarea:focus{box-shadow:0 0 0 1px #085caf;box-shadow:0 0 0 1px var(--fd-color-action-focus,#085caf);z-index:5;border-color:#085caf;border-color:var(--fd-color-action-focus,#085caf)}.fd-textarea.is-invalid{border-color:#b00;border-color:var(--fd-color-negative,#b00);border-width:2px}.fd-textarea.is-valid{border-color:#0a7e3e;border-color:var(--fd-color-positive,#0a7e3e);border-width:2px}.fd-textarea.is-alert,.fd-textarea.is-warning{border-color:#e9730c;border-color:var(--fd-color-alert,#e9730c);border-width:2px}.fd-textarea.is-disabled,.fd-textarea:disabled,.fd-textarea[aria-disabled=true]{cursor:not-allowed;color:#6a6d70;color:var(--fd-color-text-3,#6a6d70);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);background-color:#fafafa;background-color:var(--fd-color-neutral-1,#fafafa)}.fd-textarea.is-readonly,.fd-textarea[readonly]{color:#32363a;color:var(--fd-forms-color,#32363a);border-color:#eeeeef;border-color:var(--fd-color-neutral-2,#eeeeef);border-width:0 0 1px;border-radius:0}.fd-textarea--compact{height:56px;padding-top:12px}fd-input-group .fd-input-group__addon--button.fd-input-group__addon--textarea>*{max-height:100%;height:100%}.fd-input-group .fd-input-group__addon--button>*{border-color:var(--fd-forms-border-color,#89919a);height:100%;max-height:100%}"]
                }] }
    ];
    InputGroupNumberComponent.propDecorators = {
        disabled: [{ type: Input }],
        placeholder: [{ type: Input }],
        stepUpLabel: [{ type: Input }],
        stepDownLabel: [{ type: Input }]
    };
    return InputGroupNumberComponent;
}());
export { InputGroupNumberComponent };
if (false) {
    /**
     * Whether the input is disabled.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.disabled;
    /**
     * Placeholder for the input field.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.placeholder;
    /**
     * Aria label for the 'step up' button.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.stepUpLabel;
    /**
     * Aria label for the 'step down' button.
     * @type {?}
     */
    InputGroupNumberComponent.prototype.stepDownLabel;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupNumberComponent.prototype.inputTextValue;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupNumberComponent.prototype.onChange;
    /**
     * @hidden
     * @type {?}
     */
    InputGroupNumberComponent.prototype.onTouched;
}
