// @fundamental-ngx/core/esm5/lib/input-group/input-group-directives.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, forwardRef, HostBinding, Input } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
var InputGroupInputDirective = /** @class */ (function () {
    function InputGroupInputDirective() {
    }
    InputGroupInputDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-input-group-input]',
                },] }
    ];
    return InputGroupInputDirective;
}());
export { InputGroupInputDirective };
var InputGroupTextareaDirective = /** @class */ (function (_super) {
    tslib_1.__extends(InputGroupTextareaDirective, _super);
    function InputGroupTextareaDirective() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    InputGroupTextareaDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-textarea-group-input]',
                    providers: [{ provide: InputGroupInputDirective, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return InputGroupTextareaDirective; })) }]
                },] }
    ];
    return InputGroupTextareaDirective;
}(InputGroupInputDirective));
export { InputGroupTextareaDirective };
var InputGroupAddOnDirective = /** @class */ (function (_super) {
    tslib_1.__extends(InputGroupAddOnDirective, _super);
    /** @hidden */
    function InputGroupAddOnDirective(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * @hidden
         */
        _this.fdInputGroupAddonClass = true;
        /**
         * The placement of the add-on. Options include *before* and *after*
         */
        _this.placement = 'after';
        /**
         * Whether the icon add-on or the text add-on is a button.
         */
        _this.button = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    InputGroupAddOnDirective.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-input-group__addon');
        this._addClassToElement('fd-input-group__addon--' + this.placement);
        if (this.button) {
            this._addClassToElement('fd-input-group__addon--button');
        }
        if (this.type) {
            this._addClassToElement('fd-input-group__addon--' + this.type);
        }
    };
    InputGroupAddOnDirective.decorators = [
        { type: Directive, args: [{
                    // tslint:disable-next-line:directive-selector
                    selector: '[fd-input-group-addon]'
                },] }
    ];
    /** @nocollapse */
    InputGroupAddOnDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    InputGroupAddOnDirective.propDecorators = {
        fdInputGroupAddonClass: [{ type: HostBinding, args: ['class.fd-input-group__addon',] }],
        placement: [{ type: Input }],
        type: [{ type: Input }],
        button: [{ type: Input }]
    };
    return InputGroupAddOnDirective;
}(AbstractFdNgxClass));
export { InputGroupAddOnDirective };
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.fdInputGroupAddonClass;
    /**
     * The placement of the add-on. Options include *before* and *after*
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.placement;
    /**
     * The placement of the add-on. Options include *before* and *after*
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.type;
    /**
     * Whether the icon add-on or the text add-on is a button.
     * @type {?}
     */
    InputGroupAddOnDirective.prototype.button;
    /**
     * @type {?}
     * @private
     */
    InputGroupAddOnDirective.prototype.elementRef;
}
