// @smart/utils/esm2015/modules/form-builder/decorators.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/* tslint:disable:max-classes-per-file */
/**
 * @internal
 * Internal property on the constructor used for adding decorator metadata
 * so that it can be later picked up after component compilation.
 */
export const FORM_PROP = '_form_prop_';
/**
 * Base data PropDecorator.
 * @internal
 */
export class PropDecorator {
    constructor(property) {
        this.property = property;
    }
}
/**
 * @internal
 */
export class InputPropDecorator extends PropDecorator {
    constructor(property, alias) {
        super(property);
        this.alias = alias ? alias : this.property;
    }
}
/**
 * @internal
 */
export class FormPropDecorator extends PropDecorator {
}
/**
 * @internal
 */
export class ControlPropDecorator extends PropDecorator {
}
/**
 * Used for tagging dynamic inputs and adding them to the FORM_PROP property
 * of the target constructor.
 */
function makePropertyDecorator(factory) {
    return (target, key) => {
        const ctor = target.constructor;
        if (!ctor[FORM_PROP]) {
            ctor[FORM_PROP] = [];
        }
        ctor[FORM_PROP].push(factory(key));
    };
}
/**
 * Injects the AbstractForm for the dynamic form component.
 */
export function DynamicForm() {
    return makePropertyDecorator((key) => new FormPropDecorator(key));
}
/**
 * Injects the AbstractFormControl of the AbstractForm for the dynamic form component.
 */
export function DynamicFormControl() {
    return makePropertyDecorator((key) => new ControlPropDecorator(key));
}
/**
 * Injects a property of the AbstractForm for the dynamic form component.
 * Inputs are assigned from the FormSchema's 'inputs' property.
 * NOTE:
 * Property values are only available ngOnInit or onDynamicInputChange.
 * @param alias Use this alias to target a property of the AbstractForm. Defaults
 * to the assigned class property.
 * @example
 * <pre>
 *     @Component({ ... })
 *     export class DynamicFormComponent {
 *         @DynamicInput()
 *         property: string
 *     }
 * <pre>
 */
export const DynamicInput = (alias = null) => {
    return makePropertyDecorator((key) => new InputPropDecorator(key, alias));
};
