// @fundamental-ngx/core/esm5/lib/layout-grid/layout-grid.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, Input, ViewEncapsulation } from '@angular/core';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
/**
 * Use a layout grid to arrange components evenly in a grid layout.
 */
var LayoutGridComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LayoutGridComponent, _super);
    /** @hidden */
    function LayoutGridComponent(elementRef) {
        var _this = _super.call(this, elementRef) || this;
        _this.elementRef = elementRef;
        /**
         * Whether the grid should have a gap.
         */
        _this.nogap = false;
        return _this;
    }
    /** @hidden */
    /**
     * @hidden
     * @return {?}
     */
    LayoutGridComponent.prototype._setProperties = /**
     * @hidden
     * @return {?}
     */
    function () {
        this._addClassToElement('fd-layout-grid');
        if (this.nogap) {
            this._addClassToElement('fd-layout-grid--no-gap');
        }
        if (this.gapSize) {
            this._addClassToElement('fd-layout-grid--gap-' + this.gapSize);
        }
        if (this.col) {
            this._addClassToElement('fd-layout-grid--col-' + this.col);
        }
    };
    LayoutGridComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-layout-grid',
                    template: "<ng-content></ng-content>",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-layout-grid\n*/.fd-layout-grid{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:-ms-grid;display:grid;-ms-grid-columns:(1fr)[3];grid-template-columns:repeat(3,1fr);grid-gap:16px;grid-gap:var(--fd-width-gutter,16px)}.fd-layout-grid:after,.fd-layout-grid:before{box-sizing:inherit;font-size:inherit}.fd-layout-grid--no-gap{grid-gap:0}.fd-layout-grid--no-gap>*{box-shadow:none}.fd-layout-grid--gap-1{grid-gap:1px}.fd-layout-grid--gap-1>*{box-shadow:none}.fd-layout-grid--col-1{-ms-grid-columns:(1fr)[1];grid-template-columns:repeat(1,1fr)}.fd-layout-grid--col-2{-ms-grid-columns:(1fr)[2];grid-template-columns:repeat(2,1fr)}.fd-layout-grid--col-3{-ms-grid-columns:(1fr)[3];grid-template-columns:repeat(3,1fr)}.fd-layout-grid--col-4{-ms-grid-columns:(1fr)[4];grid-template-columns:repeat(4,1fr)}.fd-layout-grid--col-5{-ms-grid-columns:(1fr)[5];grid-template-columns:repeat(5,1fr)}.fd-layout-grid--col-6{-ms-grid-columns:(1fr)[6];grid-template-columns:repeat(6,1fr)}.fd-layout-grid__span-row-2{-ms-grid-row-span:2;grid-row:span 2}.fd-layout-grid__span-column-2{-ms-grid-column-span:2;grid-column:span 2}.fd-layout-grid__span-row-3{-ms-grid-row-span:3;grid-row:span 3}.fd-layout-grid__span-column-3{-ms-grid-column-span:3;grid-column:span 3}.fd-layout-grid__span-row-4{-ms-grid-row-span:4;grid-row:span 4}.fd-layout-grid__span-column-4{-ms-grid-column-span:4;grid-column:span 4}.fd-layout-grid__span-row-5{-ms-grid-row-span:5;grid-row:span 5}.fd-layout-grid__span-column-5{-ms-grid-column-span:5;grid-column:span 5}.fd-layout-grid__span-row-6{-ms-grid-row-span:6;grid-row:span 6}.fd-layout-grid__span-column-6{-ms-grid-column-span:6;grid-column:span 6}"]
                }] }
    ];
    /** @nocollapse */
    LayoutGridComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    LayoutGridComponent.propDecorators = {
        col: [{ type: Input }],
        nogap: [{ type: Input }],
        gapSize: [{ type: Input }]
    };
    return LayoutGridComponent;
}(AbstractFdNgxClass));
export { LayoutGridComponent };
if (false) {
    /**
     * \@Input Column span for the grid system
     * @type {?}
     */
    LayoutGridComponent.prototype.col;
    /**
     * Whether the grid should have a gap.
     * @type {?}
     */
    LayoutGridComponent.prototype.nogap;
    /**
     * Whether the grid should have a gap.
     * @type {?}
     */
    LayoutGridComponent.prototype.gapSize;
    /**
     * @type {?}
     * @private
     */
    LayoutGridComponent.prototype.elementRef;
}
