// @smart/utils/esm2015/services/interceptors/flaws/flaw-injection-interceptor.Module.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/* forbiddenNameSpaces useClass:false */
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { FlawInjectionInterceptor } from './flaw-injection.interceptor';
/** @internal */
let FlawInjectionInterceptorModule = class FlawInjectionInterceptorModule {
};
FlawInjectionInterceptorModule = tslib_1.__decorate([
    NgModule({
        imports: [],
        providers: [
            {
                provide: HTTP_INTERCEPTORS,
                useClass: FlawInjectionInterceptor,
                multi: true
            },
            {
                provide: APP_INITIALIZER,
                useFactory() {
                    FlawInjectionInterceptor.registerRequestFlaw({
                        test: (request) => /sites\/[\w-]+\//.test(request.url),
                        mutate: (request) => {
                            return request.clone({
                                url: request.url.replace(/sites\/([\w-]+)\//, 'sites/' + Math.random() + '/')
                            });
                        }
                    });
                    return (component) => {
                        // an initializer useFactory must return a function
                    };
                },
                multi: true
            }
        ]
    })
], FlawInjectionInterceptorModule);
export { FlawInjectionInterceptorModule };
