// @smart/utils/esm5/services/rest/rest-service.factory.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { RestClient } from './rest-client';
import { LogService } from '../log.service';
/** @internal */
var RestServiceFactory = /** @class */ (function () {
    function RestServiceFactory(httpClient) {
        this.httpClient = httpClient;
        this.map = new Map();
        this.basePath = null;
        this.DOMAIN = null;
        this.IDENTIFIER = 'identifier';
    }
    RestServiceFactory_1 = RestServiceFactory;
    RestServiceFactory.setGlobalBasePath = function (globalBasePath) {
        if (!RestServiceFactory_1.globalBasePath) {
            RestServiceFactory_1.globalBasePath = globalBasePath;
        }
        else {
            RestServiceFactory_1.logService.warn('The value of a global base path was already set. ' +
                'Update is not possible, the value remained unchanged!');
        }
    };
    RestServiceFactory.getGlobalBasePath = function () {
        return RestServiceFactory_1.globalBasePath ? RestServiceFactory_1.globalBasePath : '';
    };
    RestServiceFactory.prototype.setDomain = function (DOMAIN) {
        this.DOMAIN = DOMAIN;
    };
    RestServiceFactory.prototype.setBasePath = function (basePath) {
        this.basePath = basePath;
    };
    RestServiceFactory.prototype.get = function (uri, identifier) {
        if (identifier === void 0) { identifier = this.IDENTIFIER; }
        if (this.map.has(uri + identifier)) {
            return this.map.get(uri + identifier);
        }
        if (!/^https?\:\/\//.test(uri)) {
            var newBasePath = this.getNewBasePath();
            var basePathURI = lodash.isEmpty(newBasePath)
                ? uri
                : newBasePath + (/^\//.test(uri) ? uri : "/" + uri);
            uri = this.shouldAppendDomain(uri) ? this.DOMAIN + "/" + uri : basePathURI;
        }
        var restClient = new RestClient(this.httpClient, uri, identifier);
        this.map.set(uri + identifier, restClient);
        return restClient;
    };
    RestServiceFactory.prototype.shouldAppendDomain = function (uri) {
        return !!this.DOMAIN && !/^\//.test(uri);
    };
    RestServiceFactory.prototype.getNewBasePath = function () {
        return this.basePath ? this.basePath : RestServiceFactory_1.globalBasePath;
    };
    var RestServiceFactory_1;
    RestServiceFactory.globalBasePath = null;
    RestServiceFactory.logService = new LogService();
    RestServiceFactory.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    RestServiceFactory = RestServiceFactory_1 = tslib_1.__decorate([
        Injectable(),
        tslib_1.__metadata("design:paramtypes", [HttpClient])
    ], RestServiceFactory);
    return RestServiceFactory;
}());
export { RestServiceFactory };
