// @smart/utils/esm5/services/rest/rest-client.js
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as lodash from 'lodash';
import { HttpHeaders } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { URIBuilder } from '../../utils';
/** @internal */
var RestClient = /** @class */ (function () {
    function RestClient(httpClient, url, identifierName) {
        var _this = this;
        this.httpClient = httpClient;
        this.url = url;
        this.identifierName = identifierName;
        this.DEFAULT_HEADERS = { 'x-requested-with': 'Angular' };
        this.DEFAULT_OPTIONS = { headers: {} };
        // will activate response headers appending
        this.metadataActivated = false;
        ///////////////////////////////////////////
        /// INTERNAL METHODS NEEDED FOR GATEWAY ///
        ///////////////////////////////////////////
        this.getMethodForSingleInstance = function (name) {
            switch (name) {
                case 'getById':
                    return function (id, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.getById(id, options);
                    };
                case 'get':
                    return function (searchParams, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.get(searchParams, options);
                    };
                case 'update':
                    return function (payload, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.update(payload, options);
                    };
                case 'save':
                    return function (payload, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.save(payload, options);
                    };
                case 'patch':
                    return function (payload, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.patch(payload, options);
                    };
                case 'remove':
                    return function (payload, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.remove(payload, options);
                    };
            }
        };
        this.getMethodForArray = function (name) {
            switch (name) {
                case 'query':
                    return function (params, options) {
                        if (options === void 0) { options = _this.DEFAULT_OPTIONS; }
                        return _this.query(params, options);
                    };
            }
        };
    }
    RestClient.prototype.getById = function (identifier, options) {
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        return this.addHeadersToBody(this.httpClient.get(this.url + "/" + identifier, {
            headers: this.buildRequestHeaders(options.headers || {}),
            observe: 'response'
        })).toPromise();
    };
    RestClient.prototype.get = function (searchParams, options) {
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        var params = this.convertToTypeMapOfString(searchParams);
        return this.addHeadersToBody(this.httpClient.get(this.interpolateParamsInURL(this.url, params), {
            params: this.formatQueryString(this.determineTrueQueryStringParams(this.url, searchParams)),
            headers: this.buildRequestHeaders(options.headers || {}),
            observe: 'response'
        })).toPromise();
    };
    RestClient.prototype.query = function (searchParams, options) {
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        var params = searchParams ? this.convertToTypeMapOfString(searchParams) : searchParams;
        return this.addHeadersToBody(this.httpClient.get(this.interpolateParamsInURL(this.url, params), {
            params: this.formatQueryString(this.determineTrueQueryStringParams(this.url, searchParams)),
            headers: this.buildRequestHeaders(options.headers || {}),
            observe: 'response'
        }))
            .pipe(map(function (arr) { return arr || []; }))
            .toPromise();
    };
    RestClient.prototype.page = function (pageable, options) {
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        return (this.addHeadersToBody(this.httpClient.get(this.interpolateParamsInURL(this.url, pageable), {
            params: this.formatQueryString(this.determineTrueQueryStringParams(this.url, pageable)),
            headers: this.buildRequestHeaders(options.headers || {}),
            observe: 'response'
        }))
            // force typing to accept the fact that a page is never null
            .pipe(map(function (arr) { return arr; }))
            .toPromise());
    };
    RestClient.prototype.update = function (payload, options) {
        var _this = this;
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        return this.performIdentifierCheck(payload)
            .then(function () { return _this.buildUrlWithIdentifier(payload); })
            .then(function (url) {
            return _this.httpClient
                .put(url, payload, {
                headers: _this.buildRequestHeaders(options.headers || {})
            })
                .toPromise();
        });
    };
    RestClient.prototype.save = function (payload, options) {
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        return this.httpClient
            .post(this.interpolateParamsInURL(this.url, payload), payload, {
            headers: this.buildRequestHeaders(options.headers || {})
        })
            .toPromise();
    };
    RestClient.prototype.patch = function (payload, options) {
        var _this = this;
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        return this.performIdentifierCheck(payload)
            .then(function () { return _this.buildUrlWithIdentifier(payload); })
            .then(function (url) {
            return _this.httpClient
                .patch(url, payload, {
                headers: _this.buildRequestHeaders(options.headers || {})
            })
                .toPromise();
        });
    };
    RestClient.prototype.remove = function (payload, options) {
        var _this = this;
        if (options === void 0) { options = this.DEFAULT_OPTIONS; }
        return this.performIdentifierCheck(payload)
            .then(function () { return _this.buildUrlWithIdentifier(payload); })
            .then(function (url) {
            return _this.httpClient
                .delete(url, { headers: _this.buildRequestHeaders(options.headers || {}) })
                .toPromise();
        });
    };
    RestClient.prototype.activateMetadata = function () {
        this.metadataActivated = true;
    };
    RestClient.prototype.convertToTypeMapOfString = function (searchParams) {
        return lodash.mapValues(searchParams, function (val) { return String(val); });
    };
    RestClient.prototype.formatQueryString = function (_params) {
        return this.sortByKeys(_params);
    };
    RestClient.prototype.addHeadersToBody = function (observable) {
        var _this = this;
        return observable.pipe(map(function (response) {
            var data = response.body;
            if (_this.metadataActivated && data) {
                // used by @Cached annotation
                data.headers = response.headers;
            }
            return data;
        }));
    };
    /*
     * interpolation URL placeholders interpolation with potential matches in queryString
     */
    RestClient.prototype.interpolateParamsInURL = function (_url, payload) {
        // only keep params to be found in the URI or query params
        if (payload && typeof payload !== 'string') {
            return new URIBuilder(_url)
                .replaceParams(payload)
                .sanitize()
                .build();
        }
        else {
            return _url;
        }
    };
    /*
     * remove from queryString any param needed for URL placeholders interpolation
     */
    RestClient.prototype.determineTrueQueryStringParams = function (url, payload) {
        return typeof payload === 'object'
            ? Object.keys(payload).reduce(function (prev, next) {
                if (!new RegExp(':' + next + '/').test(url) &&
                    !new RegExp(':' + next + '$').test(url) &&
                    !new RegExp(':' + next + '&').test(url) &&
                    !lodash.isNil(payload[next])) {
                    prev[next] = payload[next];
                }
                return prev;
            }, {})
            : {};
    };
    RestClient.prototype.sortByKeys = function (obj) {
        var keys = lodash.sortBy(lodash.keys(obj), function (key) {
            return key;
        });
        return lodash.zipObject(keys, lodash.map(keys, function (key) {
            return obj[key];
        }));
    };
    RestClient.prototype.performIdentifierCheck = function (payload) {
        var identifier = typeof payload === 'string' ? payload : payload[this.identifierName];
        if (!identifier) {
            return Promise.reject('no data was found under the ' +
                identifier +
                ' field of object ' +
                JSON.stringify(payload) +
                ', it is necessary for update and remove operations');
        }
        return Promise.resolve();
    };
    RestClient.prototype.buildUrlWithIdentifier = function (payload) {
        var identifier = typeof payload === 'string' ? payload : payload[this.identifierName];
        var url = this.interpolateParamsInURL("" + this.url, payload);
        url =
            url.includes('?') || this.url.includes(':' + this.identifierName)
                ? url
                : url + "/" + identifier;
        return Promise.resolve(url);
    };
    RestClient.prototype.buildRequestHeaders = function (headers) {
        return new HttpHeaders(tslib_1.__assign({}, this.DEFAULT_HEADERS, headers));
    };
    return RestClient;
}());
export { RestClient };
