// @smart/utils/esm5/services/log.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export var LogLevel;
(function (LogLevel) {
    LogLevel[LogLevel["log"] = 0] = "log";
    LogLevel[LogLevel["debug"] = 1] = "debug";
    LogLevel[LogLevel["info"] = 2] = "info";
    LogLevel[LogLevel["warn"] = 3] = "warn";
    LogLevel[LogLevel["error"] = 4] = "error";
})(LogLevel || (LogLevel = {}));
var LogService = /** @class */ (function () {
    function LogService() {
        this.logLevel = LogLevel.info;
    }
    LogService.prototype.log = function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this._log(LogLevel.log, msg);
    };
    LogService.prototype.debug = function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this._log(LogLevel.debug, msg);
    };
    LogService.prototype.info = function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this._log(LogLevel.info, msg);
    };
    LogService.prototype.warn = function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this._log(LogLevel.warn, msg);
    };
    LogService.prototype.error = function () {
        var msg = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            msg[_i] = arguments[_i];
        }
        this._log(LogLevel.error, msg);
    };
    LogService.prototype.setLogLevel = function (logLevel) {
        this.logLevel = logLevel;
    };
    LogService.prototype._log = function (requestLevel, msg) {
        var _a;
        if (requestLevel >= this.logLevel) {
            var method = LogLevel[requestLevel];
            if (this._console() && this._console()[method]) {
                (_a = this._console())[method].apply(_a, tslib_1.__spread(msg));
            }
        }
    };
    LogService.prototype._console = function () {
        return console;
    };
    return LogService;
}());
export { LogService };
