// @smart/utils/esm5/services/browser/FingerPrintingService.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { stringUtils } from '../../utils';
/** @internal */
var FingerPrintingService = /** @class */ (function () {
    function FingerPrintingService() {
        this._fingerprint = stringUtils.encode(this.getUserAgent() +
            this.getPlugins() +
            this.hasJava() +
            this.hasFlash() +
            this.hasLocalStorage() +
            this.hasSessionStorage() +
            this.hasCookie() +
            this.getTimeZone() +
            this.getLanguage() +
            this.getSystemLanguage() +
            this.hasCanvas());
    }
    /**
     * Get unique browser fingerprint information encoded in Base64.
     */
    FingerPrintingService.prototype.getFingerprint = function () {
        return this._fingerprint;
    };
    FingerPrintingService.prototype.getUserAgent = function () {
        return navigator.userAgent;
    };
    FingerPrintingService.prototype.getPlugins = function () {
        var plugins = [];
        for (var i = 0, l = navigator.plugins.length; i < l; i++) {
            if (navigator.plugins[i]) {
                plugins.push(navigator.plugins[i].name);
            }
        }
        return plugins.join(',');
    };
    FingerPrintingService.prototype.hasJava = function () {
        return navigator.javaEnabled();
    };
    FingerPrintingService.prototype.hasFlash = function () {
        return !!navigator.plugins.namedItem('Shockwave Flash');
    };
    FingerPrintingService.prototype.hasLocalStorage = function () {
        try {
            var hasLs = 'test-has-ls';
            localStorage.setItem(hasLs, hasLs);
            localStorage.removeItem(hasLs);
            return true;
        }
        catch (exception) {
            return false;
        }
    };
    FingerPrintingService.prototype.hasSessionStorage = function () {
        try {
            var hasSs = 'test-has-ss';
            sessionStorage.setItem(hasSs, hasSs);
            sessionStorage.removeItem(hasSs);
            return true;
        }
        catch (exception) {
            return false;
        }
    };
    FingerPrintingService.prototype.hasCookie = function () {
        return navigator.cookieEnabled;
    };
    FingerPrintingService.prototype.getTimeZone = function () {
        return String(String(new Date()).split('(')[1]).split(')')[0];
    };
    FingerPrintingService.prototype.getLanguage = function () {
        return navigator.language;
    };
    FingerPrintingService.prototype.getSystemLanguage = function () {
        return window.navigator.language;
    };
    FingerPrintingService.prototype.hasCanvas = function () {
        try {
            var elem = document.createElement('canvas');
            return !!(elem.getContext && elem.getContext('2d'));
        }
        catch (e) {
            return false;
        }
    };
    return FingerPrintingService;
}());
export { FingerPrintingService };
