// @smart/utils/esm5/directives/list-keyboard-control/list-item-keyboard-control.directive.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
import { Directive, ElementRef, OnInit, Renderer2 } from '@angular/core';
/**
 * Directive for marking list item for 'ListKeyboardControlDirective' to allow for navigating with keyboard.
 */
var ListItemKeyboardControlDirective = /** @class */ (function () {
    function ListItemKeyboardControlDirective(hostElement, renderer) {
        this.hostElement = hostElement;
        this.renderer = renderer;
        /** @internal */
        this.activeClassName = 'is-active';
    }
    /** @internal */
    ListItemKeyboardControlDirective.prototype.ngOnInit = function () {
        this.setTabIndex();
    };
    ListItemKeyboardControlDirective.prototype.getElement = function () {
        return this.hostElement.nativeElement;
    };
    ListItemKeyboardControlDirective.prototype.setActive = function () {
        var elm = this.getElement();
        this.renderer.addClass(elm, this.activeClassName);
        elm.scrollIntoView({ block: 'nearest' });
    };
    ListItemKeyboardControlDirective.prototype.setInactive = function () {
        this.renderer.removeClass(this.getElement(), this.activeClassName);
    };
    /** @internal */
    ListItemKeyboardControlDirective.prototype.setTabIndex = function () {
        this.renderer.setAttribute(this.getElement(), 'tabindex', '-1');
    };
    ListItemKeyboardControlDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    ListItemKeyboardControlDirective = tslib_1.__decorate([
        Directive({
            selector: '[suListItemKeyboardControl]'
        }),
        tslib_1.__metadata("design:paramtypes", [ElementRef, Renderer2])
    ], ListItemKeyboardControlDirective);
    return ListItemKeyboardControlDirective;
}());
export { ListItemKeyboardControlDirective };
