// @smart/utils/esm5/services/interceptors/http-interceptor.module.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { APP_BOOTSTRAP_LISTENER, NgModule } from '@angular/core';
import { defaultRetryStrategyFactory, exponentialRetryStrategyFactory, linearRetryStrategyFactory, DefaultRetryStrategy, ExponentialRetry, ExponentialRetryStrategy, LinearRetry, LinearRetryStrategy, SimpleRetry } from './errors';
import { BackendInterceptor, HttpBackendService } from './backend';
import { FlawInjectionInterceptorModule } from './flaws';
import { HttpAuthInterceptor } from './http-auth.interceptor';
import { HttpErrorInterceptorService } from './http-error-interceptor.service';
import { HttpErrorInterceptor } from './http-error.interceptor';
var ɵ0 = defaultRetryStrategyFactory, ɵ1 = exponentialRetryStrategyFactory, ɵ2 = linearRetryStrategyFactory;
/**
 * @ngdoc overview
 * @name httpInterceptorModule
 *
 * @description
 * This module provides the functionality to add custom HTTP error interceptors.
 * Interceptors are used to execute code each time an HTTP request fails.
 *
 */
var HttpInterceptorModule = /** @class */ (function () {
    function HttpInterceptorModule() {
    }
    HttpInterceptorModule_1 = HttpInterceptorModule;
    HttpInterceptorModule.forRoot = function () {
        var HttpErrorInterceptorClasses = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            HttpErrorInterceptorClasses[_i] = arguments[_i];
        }
        return {
            ngModule: HttpInterceptorModule_1,
            providers: tslib_1.__spread(HttpErrorInterceptorClasses, [
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: HttpAuthInterceptor,
                    multi: true
                },
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: HttpErrorInterceptor,
                    multi: true
                },
                {
                    provide: HTTP_INTERCEPTORS,
                    useClass: BackendInterceptor,
                    multi: true
                },
                {
                    provide: APP_BOOTSTRAP_LISTENER,
                    useFactory: function (httpErrorInterceptorService) {
                        httpErrorInterceptorService.addInterceptors(HttpErrorInterceptorClasses);
                        return function (component) {
                            // an initializer useFactory must return a function
                        };
                    },
                    deps: [HttpErrorInterceptorService],
                    multi: true
                }
            ])
        };
    };
    var HttpInterceptorModule_1;
    HttpInterceptorModule = HttpInterceptorModule_1 = tslib_1.__decorate([
        NgModule({
            imports: [FlawInjectionInterceptorModule],
            providers: [
                SimpleRetry,
                LinearRetry,
                ExponentialRetry,
                {
                    provide: DefaultRetryStrategy,
                    useFactory: ɵ0,
                    deps: [SimpleRetry]
                },
                {
                    provide: ExponentialRetryStrategy,
                    useFactory: ɵ1,
                    deps: [ExponentialRetry]
                },
                {
                    provide: LinearRetryStrategy,
                    useFactory: ɵ2,
                    deps: [SimpleRetry]
                },
                HttpErrorInterceptorService,
                HttpBackendService
            ]
        })
    ], HttpInterceptorModule);
    return HttpInterceptorModule;
}());
export { HttpInterceptorModule };
export { ɵ0, ɵ1, ɵ2 };
