// @fundamental-ngx/core/esm2015/lib/notification/notification-utils/notification-ref.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Subject } from 'rxjs';
/**
 * Reference to a notification component generated via the NotificationService.
 * It can be injected into the content component through the constructor.
 * For a template, it is declared as part of the implicit context, see examples.
 */
export class NotificationRef {
    constructor() {
        this._afterClosed = new Subject();
        this._afterClosedGroup = new Subject();
        /**
         * Observable that is triggered when the notification is closed.
         * On close a *result* is passed back. On dismiss, an *error* is returned instead.
         */
        this.afterClosed = this._afterClosed.asObservable();
        /**
         * Observable that is triggered when the notification is closed.
         * On close a *result* is passed back. On dismiss, an *error* is returned instead.
         */
        this.afterClosedGroup = this._afterClosedGroup.asObservable();
    }
    /**
     * Closes the notification and passes the argument to the afterClosed observable.
     * @param {?=} result Value passed back to the observable as a result.
     * @return {?}
     */
    close(result) {
        this._afterClosed.next(result);
    }
    /**
     * Dismisses the notification and passes the argument to the afterClosed observable as an error.
     * @param {?=} reason Value passed back to the observable as an error.
     * @return {?}
     */
    dismiss(reason) {
        this._afterClosed.error(reason);
    }
    /**
     * Closes the notification group and passes the argument to the afterClosed observable
     * as an error.
     * @param {?=} reason Value passed back to the observable as an error.
     * @return {?}
     */
    closeWholeGroup(reason) {
        this._afterClosedGroup.next(reason);
    }
    /**
     * Dismisses the notification group and passes the argument to the afterClosed observable
     * as an error.
     * @param {?=} reason Value passed back to the observable as an error.
     * @return {?}
     */
    dismissWholeGroup(reason) {
        this._afterClosedGroup.error(reason);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    NotificationRef.prototype._afterClosed;
    /**
     * @type {?}
     * @private
     */
    NotificationRef.prototype._afterClosedGroup;
    /**
     * Observable that is triggered when the notification is closed.
     * On close a *result* is passed back. On dismiss, an *error* is returned instead.
     * @type {?}
     */
    NotificationRef.prototype.afterClosed;
    /**
     * Observable that is triggered when the notification is closed.
     * On close a *result* is passed back. On dismiss, an *error* is returned instead.
     * @type {?}
     */
    NotificationRef.prototype.afterClosedGroup;
    /**
     * Data passed from the calling component to the content.
     * @type {?}
     */
    NotificationRef.prototype.data;
}
