// @smart/utils/esm2015/components/select/select.component.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
var SelectComponent_1;
import * as tslib_1 from "tslib";
import { forwardRef, Component, EventEmitter, Input, Output, ViewChild, ViewEncapsulation } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { BaseValueAccessor } from '../../utils/base-value-accessor';
/**
 * @ngdoc component
 * @name  @smartutils.components:SelectComponent
 *
 * @description
 * Dropdown component allowing to select item from dropdown
 *
 * @param {ISelectItem<T>[]} items Items to displayed in dropdown
 * @param {ISelectItem<T>} initialValue Value dropdown should be initialized with
 * @param {string} placeholder String to be displayed when value is not selected
 * @param {Boolean} isKeyboardControlEnabled Flag enabling dropdown items selection with arrow keys
 * @param {Boolean} hasCustomTrigger If set true the trigger opening the dropdown will be set by user
 */
let SelectComponent = SelectComponent_1 = class SelectComponent extends BaseValueAccessor {
    /**
     * @ngdoc component
     * @name  @smartutils.components:SelectComponent
     *
     * @description
     * Dropdown component allowing to select item from dropdown
     *
     * @param {ISelectItem<T>[]} items Items to displayed in dropdown
     * @param {ISelectItem<T>} initialValue Value dropdown should be initialized with
     * @param {string} placeholder String to be displayed when value is not selected
     * @param {Boolean} isKeyboardControlEnabled Flag enabling dropdown items selection with arrow keys
     * @param {Boolean} hasCustomTrigger If set true the trigger opening the dropdown will be set by user
     */
    constructor() {
        super(...arguments);
        this.items = [];
        this.initialValue = null;
        this.placeholder = '';
        this.isKeyboardControlEnabled = true;
        this.hasCustomTrigger = false;
        this.onItemSelected = new EventEmitter();
    }
    selectItem(id) {
        const item = this.items.find((selected) => selected.id === id);
        this.setValue(item);
        this.popover.close();
        this.onItemSelected.emit(item);
    }
    ngOnChanges(changes) {
        if (changes.initialValue && changes.initialValue.currentValue && !this.value) {
            this.setInitialValue(changes.initialValue.currentValue);
        }
    }
    setInitialValue(value) {
        if (typeof value === 'number') {
            this.setValueById(value);
        }
        else {
            this.setValue(value);
        }
    }
    setValue(item) {
        this.writeValue(item);
        this.onChange(item);
    }
    setValueById(id) {
        this.setValue(this.items.find((item) => item.id === id));
    }
};
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Array)
], SelectComponent.prototype, "items", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], SelectComponent.prototype, "initialValue", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", String)
], SelectComponent.prototype, "placeholder", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SelectComponent.prototype, "isKeyboardControlEnabled", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Boolean)
], SelectComponent.prototype, "hasCustomTrigger", void 0);
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], SelectComponent.prototype, "onItemSelected", void 0);
tslib_1.__decorate([
    ViewChild('popover', { static: false }),
    tslib_1.__metadata("design:type", Object)
], SelectComponent.prototype, "popover", void 0);
SelectComponent = SelectComponent_1 = tslib_1.__decorate([
    Component({
        selector: 'su-select',
        encapsulation: ViewEncapsulation.None,
        providers: [
            { provide: NG_VALUE_ACCESSOR, useExisting: forwardRef(() => SelectComponent_1), multi: true }
        ],
        template: `
        <fd-popover #popover fillControlMode="equal" class="su-select">
            <fd-popover-control class="su-select__popover-control">
                <button
                    class="fd-dropdown__control fd-button"
                    type="button"
                    *ngIf="!hasCustomTrigger; else customTrigger"
                >
                    {{ (value && value.label) || placeholder | translate }}
                </button>

                <ng-template #customTrigger>
                    <ng-content select="[su-select-custom-trigger]"></ng-content>
                </ng-template>
            </fd-popover-control>
            <fd-popover-body>
                <fd-menu class="su-select__menu">
                    <ul
                        fd-menu-list
                        suListKeyboardControl
                        [suListKeyboardControlEnabled]="isKeyboardControlEnabled && popover.isOpen"
                        (suListKeyboardControlEnterKeydown)="selectItem($event)"
                    >
                        <li
                            fd-menu-item
                            suListItemKeyboardControl
                            [ngClass]="item.listItemClassName"
                            *ngFor="let item of items"
                            (click)="selectItem(item.id)"
                            [attr.tabindex]="-1"
                            [attr.data-select-id]="item.id"
                        >
                            {{ item.label | translate }}
                        </li>
                    </ul>
                </fd-menu>
            </fd-popover-body>
        </fd-popover>
    `,
        styles: [".su-select,.su-select__popover-control,.su-select__popover-control .fd-dropdown{width:100%}.su-select__popover-control .fd-dropdown__control{overflow:hidden;color:#51555a}.su-select__popover-control .fd-dropdown__control::after{position:absolute;top:0;right:0;display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center;padding:0;height:36px;width:36px;margin-top:0;font-size:14px}.su-select__menu .fd-menu__list{max-height:200px;overflow:auto}.su-select__menu .fd-menu__item[select-highlighted]{background:rgba(10,110,209,.07)}.su-select__menu .fd-menu__item[select-highlighted]:focus{outline:0}"]
    })
], SelectComponent);
export { SelectComponent };
