// @smart/utils/esm2015/utils/module-utils.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { APP_BOOTSTRAP_LISTENER, APP_INITIALIZER, Injectable } from '@angular/core';
/*
 * internal utility service to handle ES6 modules
 */
/* forbiddenNameSpaces angular.module:false */
/** @internal */
let ModuleUtils = class ModuleUtils {
    initialize(useFactory, deps = []) {
        return {
            provide: APP_INITIALIZER,
            useFactory() {
                useFactory.apply(undefined, Array.prototype.slice.call(arguments));
                return (component) => {
                    // an initializer useFactory must return a function
                };
            },
            deps,
            multi: true
        };
    }
    bootstrap(useFactory, deps = []) {
        return {
            provide: APP_BOOTSTRAP_LISTENER,
            useFactory() {
                useFactory.apply(undefined, Array.prototype.slice.call(arguments));
                return (component) => {
                    // an initializer useFactory must return a function
                };
            },
            deps,
            multi: true
        };
    }
    provideValues(_constants) {
        const constants = _constants || {};
        return Object.keys(constants).map((key) => {
            return {
                provide: key,
                useValue: constants[key]
            };
        });
    }
};
ModuleUtils = tslib_1.__decorate([
    Injectable()
], ModuleUtils);
export { ModuleUtils };
export const moduleUtils = new ModuleUtils();
