// @fundamental-ngx/core/esm5/lib/modal/modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, ComponentFactoryResolver, ElementRef, HostListener, Optional, TemplateRef, Type, ViewChild, ViewContainerRef, ViewEncapsulation } from '@angular/core';
import { modalFadeNgIf } from './modal-utils/modal-animations';
import { AbstractFdNgxClass } from '../utils/abstract-fd-ngx-class';
import focusTrap from 'focus-trap';
import { ModalRef } from './modal-utils/modal-ref';
var ModalComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ModalComponent, _super);
    function ModalComponent(elRef, componentFactoryResolver, cdRef, modalRef) {
        var _this = _super.call(this, elRef) || this;
        _this.elRef = elRef;
        _this.componentFactoryResolver = componentFactoryResolver;
        _this.cdRef = cdRef;
        _this.modalRef = modalRef;
        _this.escKeyCloseable = true;
        _this.focusTrapped = true;
        _this.ariaLabelledBy = null;
        _this.ariaLabel = null;
        _this.ariaDescribedBy = null;
        _this.backdropClickCloseable = true;
        _this.hasBackdrop = true;
        _this.modalPanelClass = '';
        return _this;
    }
    /**
     * @return {?}
     */
    ModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._setProperties();
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        if (this.focusTrap) {
            this.focusTrap.deactivate();
        }
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.childComponentType) {
            if (this.childComponentType instanceof Type) {
                this.loadFromComponent(this.childComponentType);
            }
            else if (this.childComponentType instanceof TemplateRef) {
                this.loadFromTemplate(this.childComponentType);
            }
        }
        if (this.focusTrapped) {
            try {
                this.focusTrap = focusTrap(this.elRef.nativeElement, {
                    clickOutsideDeactivates: this.backdropClickCloseable && this.hasBackdrop,
                    escapeDeactivates: false,
                    initialFocus: this.elRef.nativeElement
                });
                this.focusTrap.activate();
            }
            catch (e) {
                console.warn('Attempted to focus trap the modal, but no tabbable elements were found.');
            }
        }
        this.cdRef.detectChanges();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ModalComponent.prototype.closeModalEsc = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.escKeyCloseable && event.key === 'Escape') {
            this.modalRef.dismiss('escape');
        }
    };
    /**
     * @private
     * @param {?} content
     * @return {?}
     */
    ModalComponent.prototype.loadFromComponent = /**
     * @private
     * @param {?} content
     * @return {?}
     */
    function (content) {
        this.containerRef.clear();
        /** @type {?} */
        var componentFactory = this.componentFactoryResolver.resolveComponentFactory(content);
        this.componentRef = this.containerRef.createComponent(componentFactory);
    };
    /**
     * @private
     * @param {?} content
     * @return {?}
     */
    ModalComponent.prototype.loadFromTemplate = /**
     * @private
     * @param {?} content
     * @return {?}
     */
    function (content) {
        this.containerRef.clear();
        /** @type {?} */
        var context = {
            $implicit: this.modalRef
        };
        this.componentRef = this.containerRef.createEmbeddedView(content, context);
    };
    /**
     * @return {?}
     */
    ModalComponent.prototype._setProperties = /**
     * @return {?}
     */
    function () {
        if (this.modalPanelClass) {
            this._addClassToElement(this.modalPanelClass);
        }
    };
    ModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'fd-modal',
                    template: "<div class=\"fd-modal__content fd-modal__content--overrides\" role=\"document\">\n    <ng-content select=\"fd-modal-header\"></ng-content>\n    <ng-content select=\"fd-modal-body\"></ng-content>\n    <ng-content select=\"fd-modal-footer\"></ng-content>\n    <ng-container #vc></ng-container>\n    <ng-content></ng-content>\n</div>\n",
                    host: {
                        'role': 'dialog',
                        '[class.fd-modal]': 'true',
                        '[class.fd-modal-custom]': 'true',
                        '[attr.aria-labelledby]': 'ariaLabelledBy',
                        '[attr.aria-label]': 'ariaLabel',
                        '[attr.aria-describedby]': 'ariaDescribedBy',
                        '[attr.aria-modal]': 'true',
                        '[attr.id]': 'id',
                        'tabindex': '-1',
                        '[@modal-fade]': ''
                    },
                    animations: [
                        modalFadeNgIf
                    ],
                    encapsulation: ViewEncapsulation.None,
                    styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-overlay\n*/.fd-overlay{--fd-overlay-background-color:rgba(74, 80, 92, 0.3);position:fixed;display:flex;top:0;align-items:center;justify-content:center;z-index:1000}.fd-overlay.is-hidden{display:none}.fd-overlay--modal{background-color:var(--fd-overlay-background-color,rgba(74,80,92,.3));height:100vh;width:100vw}.fd-overlay--alert{position:fixed;left:50%;transform:translate(-50%);margin-top:28px}/*!\n.fd-modal\n  .fd-modal__header\n    .fd-form__title\n  .fd-form__body\n  .fd-form__footer\n*/.fd-modal{--fd-modal-background-color:var(--fd-color-background-2, #fff);--fd-modal-border-color:var(--fd-color-neutral-2, #eeeeef);--fd-modal-padding-x:var(--fd-width-gutter, 16px);font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;max-width:460px}.fd-modal:after,.fd-modal:before{box-sizing:inherit;font-size:inherit}.fd-modal__content{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;border-radius:4px;background-color:#fff;background-color:var(--fd-modal-background-color,#fff)}.fd-modal__content:after,.fd-modal__content:before{box-sizing:inherit;font-size:inherit}.fd-modal__body,.fd-modal__footer,.fd-modal__header{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:16px;margin:0;border:0;padding-left:var(--fd-modal-padding-x,16px);padding-right:var(--fd-modal-padding-x,16px)}.fd-modal__body:after,.fd-modal__body:before,.fd-modal__footer:after,.fd-modal__footer:before,.fd-modal__header:after,.fd-modal__header:before{box-sizing:inherit;font-size:inherit}.fd-modal__header{position:relative;border-bottom-color:var(--fd-modal-border-color,#eeeeef);border-bottom:1px solid #eeeeef}.fd-modal__title{text-rendering:optimizeLegibility;margin-top:0;font-size:16px;line-height:1.25;font-weight:400;margin-bottom:0;padding-right:36px}.fd-modal__title[dir=rtl],[dir=rtl] .fd-modal__title{padding-left:36px;padding-right:0}.fd-modal__close{position:absolute;right:8px;top:8px;width:36px;height:36px}.fd-modal__close[dir=rtl],[dir=rtl] .fd-modal__close{left:8px;right:auto}.fd-modal__body{max-height:calc(100vh - 128px);overflow-y:auto}.fd-modal__body>:first-child{margin-top:0}.fd-modal__body>:last-child{margin-bottom:0}.fd-modal__footer{text-align:right;border-top-color:var(--fd-modal-border-color,#eeeeef);border-top:1px solid #eeeeef}.fd-modal__footer[dir=rtl],[dir=rtl] .fd-modal__footer{text-align:left}.fd-modal-custom{max-width:none;z-index:1000;position:absolute}.fd-modal-custom:focus{outline:0}.fd-modal__content--overrides{height:100%;width:100%;min-height:inherit;min-width:inherit;max-height:inherit;max-width:inherit;display:flex;flex-direction:column}"]
                }] }
    ];
    /** @nocollapse */
    ModalComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: ComponentFactoryResolver },
        { type: ChangeDetectorRef },
        { type: ModalRef, decorators: [{ type: Optional }] }
    ]; };
    ModalComponent.propDecorators = {
        containerRef: [{ type: ViewChild, args: ['vc', { read: ViewContainerRef, static: false },] }],
        closeModalEsc: [{ type: HostListener, args: ['keyup', ['$event'],] }]
    };
    return ModalComponent;
}(AbstractFdNgxClass));
export { ModalComponent };
if (false) {
    /** @type {?} */
    ModalComponent.prototype.containerRef;
    /** @type {?} */
    ModalComponent.prototype.id;
    /** @type {?} */
    ModalComponent.prototype.escKeyCloseable;
    /** @type {?} */
    ModalComponent.prototype.focusTrapped;
    /** @type {?} */
    ModalComponent.prototype.ariaLabelledBy;
    /** @type {?} */
    ModalComponent.prototype.ariaLabel;
    /** @type {?} */
    ModalComponent.prototype.ariaDescribedBy;
    /** @type {?} */
    ModalComponent.prototype.childComponentType;
    /** @type {?} */
    ModalComponent.prototype.backdropClickCloseable;
    /** @type {?} */
    ModalComponent.prototype.hasBackdrop;
    /** @type {?} */
    ModalComponent.prototype.modalPanelClass;
    /**
     * @type {?}
     * @private
     */
    ModalComponent.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    ModalComponent.prototype.focusTrap;
    /**
     * @type {?}
     * @private
     */
    ModalComponent.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    ModalComponent.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @private
     */
    ModalComponent.prototype.cdRef;
    /**
     * @type {?}
     * @private
     */
    ModalComponent.prototype.modalRef;
}
