// @smart/utils/esm2015/services/interceptors/errors/unauthorized-error.interceptor.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { Inject, Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpEvent, HttpRequest } from '@angular/common/http';
import { IAuthenticationService, IEventService } from '../../../interfaces';
import { Deferred, HttpUtils, PromiseUtils } from '../../../utils';
import { EVENT_SERVICE, REAUTH_STARTED, WHO_AM_I_RESOURCE_URI_TOKEN } from '../../../constants';
// map used by HttpAuthInterceptor to avoid replay identical requests being held because of 401
export const GET_REQUESTS_ON_HOLD_MAP = {};
/**
 * @ngdoc service
 * @name @smartutils.services:unauthorizedErrorInterceptor
 * @description
 * Used for HTTP error code 401 (Forbidden). It will display the login modal.
 */
let UnauthorizedErrorInterceptor = class UnauthorizedErrorInterceptor {
    constructor(httpClient, authenticationService, promiseUtils, httpUtils, WHO_AM_I_RESOURCE_URI, eventService) {
        this.httpClient = httpClient;
        this.authenticationService = authenticationService;
        this.promiseUtils = promiseUtils;
        this.httpUtils = httpUtils;
        this.eventService = eventService;
        this.promisesToResolve = {}; // key: auth entry point, value: array of deferred
        this.rejectedUrls = [/authenticate/];
        this.rejectedUrls.push(WHO_AM_I_RESOURCE_URI);
    }
    predicate(request, response) {
        return (response.status === 401 &&
            (request.url
                ? this.httpUtils.isCRUDRequest(request, response) &&
                    this.isUrlNotRejected(request.url)
                : true));
    }
    responseError(request, response) {
        const deferred = this.promiseUtils.defer();
        const deferredPromise = deferred.promise.then(() => {
            return this.httpClient.request(request).toPromise();
        });
        this.authenticationService.isAuthEntryPoint(request.url).then((isAuthEntryPoint) => {
            if (!isAuthEntryPoint) {
                this.authenticationService
                    .filterEntryPoints(request.url)
                    .then((entryPoints) => {
                    const entryPoint = entryPoints[0];
                    this.promisesToResolve[entryPoint] =
                        this.promisesToResolve[entryPoint] || [];
                    this.promisesToResolve[entryPoint].push({
                        requestIdentifier: request.url,
                        deferred
                    });
                    if (this.httpUtils.isGET(request)) {
                        GET_REQUESTS_ON_HOLD_MAP[request.url] = deferredPromise;
                    }
                    this.authenticationService
                        .isReAuthInProgress(entryPoint)
                        .then((isReAuthInProgress) => {
                        if (!isReAuthInProgress) {
                            this.authenticationService
                                .setReAuthInProgress(entryPoint)
                                .then(() => {
                                const promisesToResolve = this.promisesToResolve;
                                this.eventService.publish(REAUTH_STARTED);
                                this.authenticationService
                                    .authenticate(request.url)
                                    .then(function () {
                                    promisesToResolve[this].forEach((record) => {
                                        delete GET_REQUESTS_ON_HOLD_MAP[record.requestIdentifier];
                                        record.deferred.resolve();
                                    });
                                    promisesToResolve[this] = [];
                                }.bind(entryPoint), function () {
                                    promisesToResolve[this].forEach((record) => {
                                        delete GET_REQUESTS_ON_HOLD_MAP[record.requestIdentifier];
                                        record.deferred.reject();
                                    });
                                    promisesToResolve[this] = [];
                                }.bind(entryPoint));
                            });
                        }
                    });
                });
            }
            else {
                deferred.reject(response);
            }
        });
        return deferredPromise;
    }
    isUrlNotRejected(url) {
        return !this.rejectedUrls.some((rejectedUrl) => {
            return typeof rejectedUrl === 'string'
                ? url.indexOf(rejectedUrl) === 0
                : rejectedUrl.test(url);
        });
    }
};
UnauthorizedErrorInterceptor.ctorParameters = () => [
    { type: HttpClient },
    { type: IAuthenticationService },
    { type: PromiseUtils },
    { type: HttpUtils },
    { type: String, decorators: [{ type: Inject, args: [WHO_AM_I_RESOURCE_URI_TOKEN,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [EVENT_SERVICE,] }] }
];
UnauthorizedErrorInterceptor = tslib_1.__decorate([
    Injectable(),
    tslib_1.__param(4, Inject(WHO_AM_I_RESOURCE_URI_TOKEN)),
    tslib_1.__param(5, Inject(EVENT_SERVICE)),
    tslib_1.__metadata("design:paramtypes", [HttpClient,
        IAuthenticationService,
        PromiseUtils,
        HttpUtils, String, Object])
], UnauthorizedErrorInterceptor);
export { UnauthorizedErrorInterceptor };
