// @smart/utils/esm5/utils/base-value-accessor.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
/**
 * @ngdoc object
 * @name utils.object:BaseValueAccessor
 * @description
 *
 * Class implementing {@link https://angular.io/api/forms/ControlValueAccessor ControlValueAccessor} interface used to create custom Angular inputs that
 * can be integrated with Angular Forms and.
 */
var BaseValueAccessor = /** @class */ (function () {
    function BaseValueAccessor() {
        this.disabled = false;
        this.value = null;
    }
    BaseValueAccessor.prototype.onChange = function (item) {
        // Is set by registerOnChange method
    };
    BaseValueAccessor.prototype.onTouched = function () {
        // Is set by registerOnTouched method
    };
    BaseValueAccessor.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    BaseValueAccessor.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    BaseValueAccessor.prototype.setDisabledState = function (isDisabled) {
        this.disabled = isDisabled;
    };
    BaseValueAccessor.prototype.writeValue = function (item) {
        this.value = item;
    };
    return BaseValueAccessor;
}());
export { BaseValueAccessor };
