// @smart/utils/esm5/utils/module-utils.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { APP_BOOTSTRAP_LISTENER, APP_INITIALIZER, Injectable } from '@angular/core';
/*
 * internal utility service to handle ES6 modules
 */
/* forbiddenNameSpaces angular.module:false */
/** @internal */
var ModuleUtils = /** @class */ (function () {
    function ModuleUtils() {
    }
    ModuleUtils.prototype.initialize = function (useFactory, deps) {
        if (deps === void 0) { deps = []; }
        return {
            provide: APP_INITIALIZER,
            useFactory: function () {
                useFactory.apply(undefined, Array.prototype.slice.call(arguments));
                return function (component) {
                    // an initializer useFactory must return a function
                };
            },
            deps: deps,
            multi: true
        };
    };
    ModuleUtils.prototype.bootstrap = function (useFactory, deps) {
        if (deps === void 0) { deps = []; }
        return {
            provide: APP_BOOTSTRAP_LISTENER,
            useFactory: function () {
                useFactory.apply(undefined, Array.prototype.slice.call(arguments));
                return function (component) {
                    // an initializer useFactory must return a function
                };
            },
            deps: deps,
            multi: true
        };
    };
    ModuleUtils.prototype.provideValues = function (_constants) {
        var constants = _constants || {};
        return Object.keys(constants).map(function (key) {
            return {
                provide: key,
                useValue: constants[key]
            };
        });
    };
    ModuleUtils = tslib_1.__decorate([
        Injectable()
    ], ModuleUtils);
    return ModuleUtils;
}());
export { ModuleUtils };
export var moduleUtils = new ModuleUtils();
