// @smart/utils/esm5/utils/window-utils.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { commonNgZone } from './common-ng-zone';
/**
 * @ngdoc service
 * @name @smartutils.services:WindowUtils
 *
 * @description
 * A collection of utility methods for windows.
 */
var WindowUtils = /** @class */ (function () {
    function WindowUtils(ngZone) {
        var _this = this;
        this.ngZone = ngZone;
        /**
         * @ngdoc method
         * @name @smartutils.services:WindowUtils#isIframe
         * @methodOf @smartutils.services:WindowUtils
         * @description
         * <b>isIframe</b> will check if the current document is in an iFrame.
         * @returns {boolean} true if the current document is in an iFrame.
         */
        this.isIframe = function () {
            return _this.getWindow().top !== _this.getWindow();
        };
        this.ngZone = this.ngZone || commonNgZone;
    }
    WindowUtils.prototype.getWindow = function () {
        return window;
    };
    /**
     * @ngdoc method
     * @name @smartutils.services:WindowUtils#runTimeoutOutsideAngular
     * @methodOf @smartutils.services:WindowUtils
     *
     * @description
     * Runs a given timeout outside Angular and attaches its callback to Angular
     * this is usefull in order not to be blocking from an e2e stand point
     *
     * @param {string} callback argument less callback to execute when timeout.
     * @param {number} timeout the delay in milliseconds until timeout
     */
    WindowUtils.prototype.runTimeoutOutsideAngular = function (callback, timeout) {
        var ngZone = this.ngZone;
        if (ngZone !== undefined) {
            return ngZone.runOutsideAngular(function () {
                return setTimeout(function () {
                    return ngZone.run(callback);
                }, timeout);
            });
        }
        else {
            throw new Error('this instance of WindowUtils has not been instantiated through Angular 7 DI');
        }
    };
    /**
     * @ngdoc method
     * @name @smartutils.services:WindowUtils#runIntervalOutsideAngular
     * @methodOf @smartutils.services:WindowUtils
     *
     * @description
     * Runs a given interval outside Angular and attaches its callback to Angular
     * this is usefull in order not to be blocking from an e2e stand point
     *
     * @param {string} callback argument less callback to execute when timeout.
     * @param {number} timeout the delay in milliseconds until timeout
     */
    WindowUtils.prototype.runIntervalOutsideAngular = function (callback, timeout) {
        var ngZone = this.ngZone;
        if (ngZone === undefined) {
            throw new Error('this instance of WindowUtils has not been instantiated through Angular 7 DI');
        }
        return ngZone.runOutsideAngular(function () {
            return setInterval(function () {
                return ngZone.run(callback);
            }, timeout);
        });
    };
    WindowUtils.SMARTEDIT_IFRAME_ID = 'ySmartEditFrame';
    return WindowUtils;
}());
export { WindowUtils };
export var windowUtils = new WindowUtils();
