// @smart/utils/esm2015/modules/form-builder/components/form-lister/form-lister.component.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import * as tslib_1 from "tslib";
/**
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
import { Component } from '@angular/core';
import { values } from 'lodash';
import { FormGrouping } from '../../models';
import { DynamicForm } from '../../decorators';
/**
 * Default component used for listing forms for a group form.
 */
let FormListerComponent = class FormListerComponent {
    get forms() {
        return values(this.form.forms);
    }
};
tslib_1.__decorate([
    DynamicForm(),
    tslib_1.__metadata("design:type", FormGrouping)
], FormListerComponent.prototype, "form", void 0);
FormListerComponent = tslib_1.__decorate([
    Component({
        selector: 'form-lister',
        template: "<ng-template [formRenderer]=\"form\" *ngFor=\"let form of forms\"></ng-template>\n",
        styles: [`
            :host {
                display: block;
            }
        `]
    })
], FormListerComponent);
export { FormListerComponent };
