// @smart/utils/esm5/interfaces/i-alert.service.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export var IAlertServiceType;
(function (IAlertServiceType) {
    IAlertServiceType["INFO"] = "information";
    IAlertServiceType["SUCCESS"] = "success";
    IAlertServiceType["WARNING"] = "warning";
    IAlertServiceType["DANGER"] = "error";
})(IAlertServiceType || (IAlertServiceType = {}));
var IAlertService = /** @class */ (function () {
    function IAlertService() {
    }
    IAlertService.prototype.showAlert = function (alertConf) {
        'proxyFunction';
        return;
    };
    IAlertService.prototype.showInfo = function (alertConf) {
        'proxyFunction';
        return;
    };
    IAlertService.prototype.showDanger = function (alertConf) {
        'proxyFunction';
        return;
    };
    IAlertService.prototype.showWarning = function (alertConf) {
        'proxyFunction';
        return;
    };
    IAlertService.prototype.showSuccess = function (alertConf) {
        'proxyFunction';
        return;
    };
    return IAlertService;
}());
export { IAlertService };
