// @smart/utils/esm5/interfaces/i-modal-button-options.js
/**
 * Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved.
 * @module smartutils
 */
export var FundamentalModalButtonStyle;
(function (FundamentalModalButtonStyle) {
    FundamentalModalButtonStyle["Default"] = "light";
    FundamentalModalButtonStyle["Primary"] = "emphasized";
})(FundamentalModalButtonStyle || (FundamentalModalButtonStyle = {}));
export var FundamentalModalButtonAction;
(function (FundamentalModalButtonAction) {
    FundamentalModalButtonAction["Close"] = "close";
    FundamentalModalButtonAction["Dismiss"] = "dismiss";
    FundamentalModalButtonAction["None"] = "none";
})(FundamentalModalButtonAction || (FundamentalModalButtonAction = {}));
/**
 * @ngdoc object
 * @name modalServiceModule.object:ModalButtonActions
 * @description
 * An enum type representing buttons available actions.
 */
export var ModalButtonActions;
(function (ModalButtonActions) {
    /**
     * @ngdoc property
     * @name None
     * @propertyOf modalServiceModule.object:ModalButtonActions
     * @description none
     *
     */
    ModalButtonActions["None"] = "none";
    /**
     * @ngdoc property
     * @name Close
     * @propertyOf modalServiceModule.object:ModalButtonActions
     * @description close
     *
     */
    ModalButtonActions["Close"] = "close";
    /**
     * @ngdoc property
     * @name Dismiss
     * @propertyOf modalServiceModule.object:ModalButtonActions
     * @description dismiss
     *
     */
    ModalButtonActions["Dismiss"] = "dismiss";
})(ModalButtonActions || (ModalButtonActions = {}));
/**
 * @ngdoc object
 * @name modalServiceModule.object:ModalButtonStyles
 * @description
 * An enum type representing buttons available styles.
 */
export var ModalButtonStyles;
(function (ModalButtonStyles) {
    /**
     * @ngdoc property
     * @name Default
     * @propertyOf modalServiceModule.object:ModalButtonStyles
     * @description default
     */
    ModalButtonStyles["Default"] = "default";
    /**
     * @ngdoc property
     * @name Primary
     * @propertyOf modalServiceModule.object:ModalButtonStyles
     * @description primary
     */
    ModalButtonStyles["Primary"] = "primary";
})(ModalButtonStyles || (ModalButtonStyles = {}));
