// @fundamental-ngx/core/esm2015/lib/shellbar/product-menu/product-menu.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChild, HostListener, Input, ViewChild, ViewEncapsulation } from '@angular/core';
import { PopoverComponent } from '../../popover/popover.component';
import { MenuComponent } from '../../menu/menu.component';
/**
 * The component that represents a product menu.
 * Product menu is used for navigating to applications within the product.
 * ```html
 * <fd-product-menu [control]="productMenuControl"
 *                  [items]="productMenuItems">
 * </fd-product-menu>
 * ```
 */
export class ProductMenuComponent extends PopoverComponent {
    constructor() {
        super(...arguments);
        /**
         * @hidden
         */
        this.productMenuCollapsed = false;
        /**
         * When set to true, popover list will be closed after selecting the option
         */
        this.closePopoverOnSelect = false;
    }
    /**
     * @hidden
     * @return {?}
     */
    onResize() {
        /** @type {?} */
        const mq = window.matchMedia('(max-width: 601px)');
        mq.matches ? this.productMenuCollapsed = true : this.productMenuCollapsed = false;
    }
    /**
     * @hidden
     * @return {?}
     */
    ngOnInit() {
        this.onResize();
    }
    /**
     * @hidden
     * @param {?} item
     * @param {?} event
     * @return {?}
     */
    itemClicked(item, event) {
        if (this.closePopoverOnSelect) {
            this.popoverComponent.close();
        }
        item.callback(event);
    }
    /**
     * @hidden
     * @return {?}
     */
    isAnyGlyphInItems() {
        if (!this.items || this.items.length === 0) {
            return false;
        }
        return !!this.items.find((/**
         * @param {?} item
         * @return {?}
         */
        item => item.glyph));
    }
}
ProductMenuComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-product-menu',
                template: "<div class=\"fd-product-menu\">\n    <ng-template [ngIf]=\"items?.length > 0\" [ngIfElse]=\"subHeader\">\n        <fd-popover [options]=\"{placement: 'bottom-end'}\"\n                    [disabled]=\"!items || items.length === 0 || disabled\"\n                    [closeOnEscapeKey]=\"closeOnEscapeKey\"\n                    [closeOnOutsideClick]=\"closeOnOutsideClick\"\n                    [(isOpen)]=\"isOpen\"\n                    (isOpenChange)=\"openChanged($event)\"\n                    [isDropdown]=\"isDropdown\"\n                    [triggers]=\"triggers\"\n                    [focusTrapped]=\"false\">\n            <fd-popover-control>\n                <button class=\"fd-product-menu__control\">\n                    <span class=\"fd-shellbar__title fd-product-menu__title\">\n                        {{control}}\n                    </span>\n                </button>\n            </fd-popover-control>\n            <fd-popover-body>\n                <fd-menu fd-menu-addon [fdMenuAddonClass]=\"isAnyGlyphInItems()\" *ngIf=\"!menuComponent\">\n                    <ul fd-menu-list>\n                        <ng-container *ngFor=\"let item of items\">\n                            <li (click)=\"itemClicked(item, $event)\" *ngIf=\"item.link\">\n                                <a href=\"{{item.link}}\"\n                                   fd-menu-item\n                                   [ngClass]=\"{'fd-product-menu-padding-left': !item.glyph}\"\n                                   class=\"fd-product-menu-padding-right\">\n                                    {{item.name}}\n                                    <div fd-menu-item-addon *ngIf=\"item.glyph\">\n                                        <fd-icon [glyph]=\"item.glyph\"></fd-icon>\n                                    </div>\n                                </a>\n                            </li>\n                            <li *ngIf=\"!item.link\"\n                                fd-menu-item\n                                (click)=\"itemClicked(item, $event)\"\n                                [ngClass]=\"{'fd-product-menu-padding-left': !item.glyph}\"\n                                class=\"fd-product-menu-padding-right\">\n                                {{item.name}}\n                                <div fd-menu-item-addon *ngIf=\"item.glyph\">\n                                    <fd-icon [glyph]=\"item.glyph\"></fd-icon>\n                                </div>\n                            </li>\n                        </ng-container>\n                    </ul>\n                </fd-menu>\n                <ng-content select=\"fd-menu\"></ng-content>\n            </fd-popover-body>\n        </fd-popover>\n    </ng-template>\n</div>\n\n\n<ng-template #subHeader>\n    <span class=\"fd-shellbar__title fd-product-menu__title\">\n        {{control}}\n    </span>\n</ng-template>\n",
                encapsulation: ViewEncapsulation.None
            }] }
];
ProductMenuComponent.propDecorators = {
    popoverComponent: [{ type: ViewChild, args: [PopoverComponent, { static: false },] }],
    menuComponent: [{ type: ContentChild, args: [MenuComponent, { static: false },] }],
    control: [{ type: Input }],
    items: [{ type: Input }],
    closePopoverOnSelect: [{ type: Input }],
    onResize: [{ type: HostListener, args: ['window:resize', [],] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ProductMenuComponent.prototype.popoverComponent;
    /**
     * @hidden
     * @type {?}
     */
    ProductMenuComponent.prototype.menuComponent;
    /**
     * The control element to toggle the product menu,
     * represented by the name of the current application.
     * @type {?}
     */
    ProductMenuComponent.prototype.control;
    /**
     *  The items in the product menu. If the value is not passed or there is empty array,
     * the template changes and shows only control
     * @type {?}
     */
    ProductMenuComponent.prototype.items;
    /**
     * @hidden
     * @type {?}
     */
    ProductMenuComponent.prototype.productMenuCollapsed;
    /**
     * When set to true, popover list will be closed after selecting the option
     * @type {?}
     */
    ProductMenuComponent.prototype.closePopoverOnSelect;
}
