// @fundamental-ngx/core/esm2015/lib/button-group/button-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewEncapsulation } from '@angular/core';
/**
 * Container for grouped buttons.
 *
 * ```html
 * <fd-button-group>
 *     <button fd-button-grouped>Button</button>
 * </fd-button-group>
 * ```
 */
export class ButtonGroupComponent {
    constructor() {
        /**
         * @hidden
         */
        this.fdButtonGroupClass = true;
    }
}
ButtonGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'fd-button-group',
                template: "<ng-content></ng-content>\n",
                host: {
                    'role': 'group'
                },
                encapsulation: ViewEncapsulation.None,
                styles: ["/*!\n * Fundamental Styles v0.3.0\n * Copyright (c) 2019 SAP SE or an SAP affiliate company.\n * Licensed under Apache License 2.0 (https://github.com/SAP/fundamental-styles/blob/master/LICENSE)\n *//*!\n.fd-button-group+()\n    .fd-button\n*/.fd-button-group{font-size:14px;line-height:1.42857;color:#32363a;color:var(--fd-color-text-1,#32363a);font-family:\"72\",-apple-system,BlinkMacSystemFont,Segoe UI,Roboto,Helvetica Neue,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;font-weight:400;-webkit-font-smoothing:antialiased;box-sizing:border-box;padding:0;margin:0;border:0;display:inline-flex;vertical-align:middle}.fd-button-group:after,.fd-button-group:before{box-sizing:inherit;font-size:inherit}.fd-button-group>*{margin:0}.fd-button-group>:not(:first-child):not(:last-child){border-radius:0}.fd-button-group>:not(:last-child){border-right-width:0}.fd-button-group>:first-child{border-top-right-radius:0;border-bottom-right-radius:0}.fd-button-group>:last-child{border-top-left-radius:0;border-bottom-left-radius:0}.fd-button-group>[dir=rtl]:not(:first-child),[dir=rtl] .fd-button-group>:not(:first-child){border-right:none}.fd-button-group>[dir=rtl]:last-child,[dir=rtl] .fd-button-group>:last-child{border-radius:4px 0 0 4px}.fd-button-group>[dir=rtl]:first-child,[dir=rtl] .fd-button-group>:first-child{border-right-width:1px;border-radius:0 4px 4px 0}"]
            }] }
];
ButtonGroupComponent.propDecorators = {
    fdButtonGroupClass: [{ type: HostBinding, args: ['class.fd-button-group',] }]
};
if (false) {
    /**
     * @hidden
     * @type {?}
     */
    ButtonGroupComponent.prototype.fdButtonGroupClass;
}
